// @tatumio/tatum/dist/src/model/validation/Mint721BuiltInPrivateKeyValidator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mint721BuiltInPrivateKeyValidator = void 0;
const class_validator_1 = require("class-validator");
const request_1 = require("../request");
const SignatureIdValidator_1 = require("./SignatureIdValidator");
let Mint721BuiltInPrivateKeyValidator = class Mint721BuiltInPrivateKeyValidator {
    constructor() {
        this.message = null;
    }
    defaultMessage(validationArguments) {
        if (this.message) {
            return this.message;
        }
        return 'If you fill signatureId or privateKey/secret/fromPrivateKey, then tokenId, contractAddress must be present.';
    }
    validate(value, validationArguments) {
        const data = validationArguments === null || validationArguments === void 0 ? void 0 : validationArguments.object;
        const chains = [request_1.Currency.BSC, request_1.Currency.ETH, request_1.Currency.CELO, request_1.Currency.ONE, request_1.Currency.MATIC, request_1.Currency.KLAY, request_1.Currency.ALGO, request_1.Currency.SOL];
        const isAllowedChain = chains.includes(data.chain);
        if (data.minter) {
            if (data.fromPrivateKey || data.signatureId) {
                this.message = `If minter is present, no private key or signatureId must be entered. Minter is available only on these chains - ${chains}`;
                return false;
            }
            if (!this.validateRequiredFields(data)) {
                return false;
            }
        }
        if (!data.fromPrivateKey && !data.signatureId) {
            if (isAllowedChain) {
                return true;
            }
            else {
                if (!this.validateNonBuiltInPrivateKey(data, validationArguments)) {
                    return false;
                }
            }
        }
        else {
            if (!this.validateNonBuiltInPrivateKey(data, validationArguments)) {
                return false;
            }
        }
        return true;
    }
    validateNonBuiltInPrivateKey(data, validationArguments) {
        if (!this.validateRequiredFields(data)) {
            return false;
        }
        if (!data.fromPrivateKey && !data.signatureId) {
            this.message = 'Field fromPrivateKey or signatureId must be filled.';
            return false;
        }
        const signatureIdValidation = new SignatureIdValidator_1.SignatureIdValidator();
        return signatureIdValidation.validate(data, validationArguments);
    }
    validateRequiredFields(data) {
        if (data.chain === request_1.Currency.SOL) {
            return true;
        }
        if (data.chain === request_1.Currency.CELO && (!data.feeCurrency || ![request_1.Currency.CELO, request_1.Currency.CUSD, request_1.Currency.CEUR].includes(data.feeCurrency))) {
            this.message = 'CELO chain must have assigned feeCurrency field.';
            return false;
        }
        if (!data.tokenId && data.chain !== request_1.Currency.ALGO) {
            this.message = 'Field tokenId must be present.';
            return false;
        }
        if (!data.contractAddress && data.chain !== request_1.Currency.ALGO) {
            this.message = 'Field contractAddress must be present.';
            return false;
        }
        return true;
    }
};
Mint721BuiltInPrivateKeyValidator = __decorate([
    class_validator_1.ValidatorConstraint({ name: 'builtInPrivateKey', async: false })
], Mint721BuiltInPrivateKeyValidator);
exports.Mint721BuiltInPrivateKeyValidator = Mint721BuiltInPrivateKeyValidator;
