// @tatumio/tatum/dist/src/model/validation/OneOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneOf = void 0;
const class_validator_1 = require("class-validator");
// eslint-disable-next-line @typescript-eslint/ban-types
const OneOf = (oneOfFields, validationOptions) => (object, propertyName) => {
    class_validator_1.registerDecorator({
        name: 'oneOf',
        target: object.constructor,
        propertyName,
        constraints: [oneOfFields],
        options: validationOptions,
        validator: {
            validate(value, args) {
                let isSet = false;
                for (const oneOfField in args.object) {
                    if (oneOfFields.includes(oneOfField)) {
                        // @ts-ignore
                        if (args.object[oneOfField]) {
                            if (isSet) {
                                return false;
                            }
                            isSet = true;
                        }
                    }
                }
                return isSet;
            },
            defaultMessage: class_validator_1.buildMessage(() => `Exactly one of ${oneOfFields.join(", ")} fields can be filled.`, validationOptions),
        },
    });
};
exports.OneOf = OneOf;
