// @tatumio/tatum/dist/src/blockchain/dogecoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dogeGetTransaction = exports.dogeGetUTXO = exports.dogeGetBlockHash = exports.dogeGetBlock = exports.dogeGetCurrentBlock = exports.dogeBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeBroadcast" target="_blank">Tatum API documentation</a>
 */
const dogeBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/dogecoin/broadcast`, { txData, signatureId });
exports.dogeBroadcast = dogeBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
const dogeGetCurrentBlock = async () => tatum_1.get(`/v3/dogecoin/info`);
exports.dogeGetCurrentBlock = dogeGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlock" target="_blank">Tatum API documentation</a>
 */
const dogeGetBlock = async (hash) => tatum_1.get(`/v3/dogecoin/block/${hash}`);
exports.dogeGetBlock = dogeGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlockHash" target="_blank">Tatum API documentation</a>
 */
const dogeGetBlockHash = async (i) => tatum_1.get(`/v3/dogecoin/block/hash/${i}`);
exports.dogeGetBlockHash = dogeGetBlockHash;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetUTXO" target="_blank">Tatum API documentation</a>
 */
const dogeGetUTXO = async (hash, i) => tatum_1.get(`/v3/dogecoin/utxo/${hash}/${i}`);
exports.dogeGetUTXO = dogeGetUTXO;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
const dogeGetTransaction = async (hash) => tatum_1.get(`/v3/dogecoin/transaction/${hash}`);
exports.dogeGetTransaction = dogeGetTransaction;
