// @tatumio/tatum/dist/src/blockchain/egld.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.egldEstimateGas = exports.egldGetTransaction = exports.egldGetAccountErc20Balance = exports.egldGetAccountBalance = exports.egldGetBlock = exports.egldGetCurrentBlock = exports.egldGetTransactionsCount = exports.egldBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldBroadcast" target="_blank">Tatum API documentation</a>
 */
const egldBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/egld/broadcast`, { txData, signatureId });
exports.egldBroadcast = egldBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
const egldGetTransactionsCount = async (address) => tatum_1.get(`/v3/egld/transaction/count/${address}`);
exports.egldGetTransactionsCount = egldGetTransactionsCount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const egldGetCurrentBlock = async () => tatum_1.get(`/v3/egld/block/current`);
exports.egldGetCurrentBlock = egldGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldGetBlock" target="_blank">Tatum API documentation</a>
 */
const egldGetBlock = async (hash) => tatum_1.get(`/v3/egld/block/${hash}`);
exports.egldGetBlock = egldGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldGetBalance" target="_blank">Tatum API documentation</a>
 */
const egldGetAccountBalance = async (address) => tatum_1.get(`/v3/egld/account/balance/${address}`);
exports.egldGetAccountBalance = egldGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc" target="_blank">Tatum API documentation</a>
 */
const egldGetAccountErc20Balance = async (address, tokenId) => tatum_1.get(`/v3/egld/account/esdt/balance/${address}/${tokenId}`);
exports.egldGetAccountErc20Balance = egldGetAccountErc20Balance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldGetTransaction" target="_blank">Tatum API documentation</a>
 */
const egldGetTransaction = async (hash) => tatum_1.get(`/v3/egld/transaction/${hash}`);
exports.egldGetTransaction = egldGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EgldEstimateGas" target="_blank">Tatum API documentation</a>
 */
const egldEstimateGas = (body) => tatum_1.post('/v3/egld/gas', body, model_1.EgldSendTransaction);
exports.egldEstimateGas = egldEstimateGas;
