// @tatumio/tatum/dist/src/blockchain/ethereum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ethEstimateGas = exports.ethGetAccountTransactions = exports.ethGetTransaction = exports.ethGetAccountErc20Address = exports.ethGetAccountBalance = exports.ethGetBlock = exports.ethGetCurrentBlock = exports.ethGetTransactionsCount = exports.ethBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthBroadcast" target="_blank">Tatum API documentation</a>
 */
const ethBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/ethereum/broadcast`, { txData, signatureId });
exports.ethBroadcast = ethBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
const ethGetTransactionsCount = async (address) => tatum_1.get(`/v3/ethereum/transaction/count/${address}`);
exports.ethGetTransactionsCount = ethGetTransactionsCount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const ethGetCurrentBlock = async () => tatum_1.get(`/v3/ethereum/block/current`);
exports.ethGetCurrentBlock = ethGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetBlock" target="_blank">Tatum API documentation</a>
 */
const ethGetBlock = async (hash) => tatum_1.get(`/v3/ethereum/block/${hash}`);
exports.ethGetBlock = ethGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetBalance" target="_blank">Tatum API documentation</a>
 */
const ethGetAccountBalance = async (address) => tatum_1.get(`/v3/ethereum/account/balance/${address}`);
exports.ethGetAccountBalance = ethGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthErc20GetBalance" target="_blank">Tatum API documentation</a>
 */
const ethGetAccountErc20Address = async (address, contractAddress) => tatum_1.get(`/v3/ethereum/account/balance/erc20/${address}?contractAddress=${contractAddress}`);
exports.ethGetAccountErc20Address = ethGetAccountErc20Address;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransaction" target="_blank">Tatum API documentation</a>
 */
const ethGetTransaction = async (hash) => tatum_1.get(`/v3/ethereum/transaction/${hash}`);
exports.ethGetTransaction = ethGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransactionByAddress" target="_blank">Tatum API documentation</a>
 */
const ethGetAccountTransactions = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ethereum/account/transaction/${address}?pageSize=${pageSize}&offset=${offset}`);
exports.ethGetAccountTransactions = ethGetAccountTransactions;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthEstimateGas" target="_blank">Tatum API documentation</a>
 */
const ethEstimateGas = (body) => tatum_1.post('/v3/ethereum/gas', body, model_1.EstimateGasEth);
exports.ethEstimateGas = ethEstimateGas;
