// @tatumio/tatum/dist/src/blockchain/flow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flowGetTransaction = exports.flowGetAccount = exports.flowGetBlock = exports.flowBroadcastTx = exports.flowGetCurrentBlock = exports.flowGetSignKey = exports.flowSignWithKey = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowSign" target="_blank">Tatum API documentation</a>
 */
const flowSignWithKey = async (data, isPayer) => tatum_1.post('/v3/flow/sign', { data, isPayer });
exports.flowSignWithKey = flowSignWithKey;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowGetSignKey" target="_blank">Tatum API documentation</a>
 */
const flowGetSignKey = async (isPayer) => tatum_1.get(`/v3/flow/proposal/${isPayer}`);
exports.flowGetSignKey = flowGetSignKey;
// TODO: return types for FLOW
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const flowGetCurrentBlock = async () => tatum_1.get(`/v3/flow/block/current`);
exports.flowGetCurrentBlock = flowGetCurrentBlock;
const flowBroadcastTx = async (txData, signatureId, proposalKey) => tatum_1.post('/v3/flow/broadcast', { txData, signatureId, proposalKey });
exports.flowBroadcastTx = flowBroadcastTx;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowGetBlock" target="_blank">Tatum API documentation</a>
 */
const flowGetBlock = async (hash) => tatum_1.get(`/v3/flow/block/${hash}`);
exports.flowGetBlock = flowGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowGetAccount" target="_blank">Tatum API documentation</a>
 */
const flowGetAccount = async (address) => tatum_1.get(`/v3/flow/account/${address}`);
exports.flowGetAccount = flowGetAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/FlowGetTransaction" target="_blank">Tatum API documentation</a>
 */
const flowGetTransaction = async (hash) => tatum_1.get(`/v3/flow/transaction/${hash}`);
exports.flowGetTransaction = flowGetTransaction;
