// @tatumio/tatum/dist/src/blockchain/klay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.klaytnEstimateGas = exports.klaytnGetTransaction = exports.klaytnGetAccountBalance = exports.klaytnGetBlock = exports.klaytnGetCurrentBlock = exports.klaytnGetTransactionsCount = exports.klaytnBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayBroadcast" target="_blank">Tatum API documentation</a>
 */
const klaytnBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/klaytn/broadcast`, { txData, signatureId });
exports.klaytnBroadcast = klaytnBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
const klaytnGetTransactionsCount = async (address) => tatum_1.get(`/v3/klaytn/transaction/count/${address}`);
exports.klaytnGetTransactionsCount = klaytnGetTransactionsCount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const klaytnGetCurrentBlock = async () => tatum_1.get(`/v3/klaytn/block/current`);
exports.klaytnGetCurrentBlock = klaytnGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayGetBlock" target="_blank">Tatum API documentation</a>
 */
const klaytnGetBlock = async (hash) => tatum_1.get(`/v3/klaytn/block/${hash}`);
exports.klaytnGetBlock = klaytnGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayGetBalance" target="_blank">Tatum API documentation</a>
 */
const klaytnGetAccountBalance = async (address) => tatum_1.get(`/v3/klaytn/account/balance/${address}`);
exports.klaytnGetAccountBalance = klaytnGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayGetTransaction" target="_blank">Tatum API documentation</a>
 */
const klaytnGetTransaction = async (hash) => tatum_1.get(`/v3/klaytn/transaction/${hash}`);
exports.klaytnGetTransaction = klaytnGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/KlayEstimateGas" target="_blank">Tatum API documentation</a>
 */
const klaytnEstimateGas = (body) => tatum_1.post('/v3/klaytn/gas', body, model_1.EstimateGasEth);
exports.klaytnEstimateGas = klaytnEstimateGas;
