// @tatumio/tatum/dist/src/blockchain/litecoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ltcGetTransaction = exports.ltcGetTxForAccount = exports.ltcGetUTXO = exports.ltcGetBlockHash = exports.ltcGetBlock = exports.ltcGetCurrentBlock = exports.ltcGetBalance = exports.ltcBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcBroadcast" target="_blank">Tatum API documentation</a>
 */
const ltcBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/litecoin/broadcast`, { txData, signatureId });
exports.ltcBroadcast = ltcBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBalanceOfAddress" target="_blank">Tatum API documentation</a>
 */
const ltcGetBalance = async (address) => tatum_1.get(`/v3/litecoin/address/balance/${address}`);
exports.ltcGetBalance = ltcGetBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
const ltcGetCurrentBlock = async () => tatum_1.get(`/v3/litecoin/info`);
exports.ltcGetCurrentBlock = ltcGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlock" target="_blank">Tatum API documentation</a>
 */
const ltcGetBlock = async (hash) => tatum_1.get(`/v3/litecoin/block/${hash}`);
exports.ltcGetBlock = ltcGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlockHash" target="_blank">Tatum API documentation</a>
 */
const ltcGetBlockHash = async (i) => tatum_1.get(`/v3/litecoin/block/hash/${i}`);
exports.ltcGetBlockHash = ltcGetBlockHash;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetUTXO" target="_blank">Tatum API documentation</a>
 */
const ltcGetUTXO = async (hash, i) => tatum_1.get(`/v3/litecoin/utxo/${hash}/${i}`);
exports.ltcGetUTXO = ltcGetUTXO;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
const ltcGetTxForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/litecoin/transaction/address/${address}?pageSize=${pageSize}&offset=${offset}`);
exports.ltcGetTxForAccount = ltcGetTxForAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
const ltcGetTransaction = async (hash) => tatum_1.get(`/v3/litecoin/transaction/${hash}`);
exports.ltcGetTransaction = ltcGetTransaction;
