// @tatumio/tatum/dist/src/blockchain/one.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oneGetTransactionCount = exports.oneGetTransaction = exports.oneGetBalance = exports.oneGetBlock = exports.oneGetCurrentBlock = exports.oneBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneBroadcast" target="_blank">Tatum API documentation</a>
 */
const oneBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/one/broadcast`, { txData, signatureId });
exports.oneBroadcast = oneBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const oneGetCurrentBlock = async () => tatum_1.get(`/v3/one/block/current`);
exports.oneGetCurrentBlock = oneGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneGetBlock" target="_blank">Tatum API documentation</a>
 */
const oneGetBlock = async (hash) => tatum_1.get(`/v3/one/block/${hash}`);
exports.oneGetBlock = oneGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneGetBalance" target="_blank">Tatum API documentation</a>
 */
const oneGetBalance = async (address) => tatum_1.get(`/v3/one/account/balance/${address}`);
exports.oneGetBalance = oneGetBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneGetTransaction" target="_blank">Tatum API documentation</a>
 */
const oneGetTransaction = async (hash) => tatum_1.get(`/v3/one/transaction/${hash}`);
exports.oneGetTransaction = oneGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/OneGetBalance" target="_blank">Tatum API documentation</a>
 */
const oneGetTransactionCount = async (address) => tatum_1.get(`/v3/one/transaction/count/${address}`);
exports.oneGetTransactionCount = oneGetTransactionCount;
