// @tatumio/tatum/dist/src/blockchain/polygon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.polygonEstimateGas = exports.polygonGetTransaction = exports.polygonGetAccountBalance = exports.polygonGetBlock = exports.polygonGetCurrentBlock = exports.polygonGetTransactionsCount = exports.polygonBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonBroadcast" target="_blank">Tatum API documentation</a>
 */
const polygonBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/polygon/broadcast`, { txData, signatureId });
exports.polygonBroadcast = polygonBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
const polygonGetTransactionsCount = async (address) => tatum_1.get(`/v3/polygon/transaction/count/${address}`);
exports.polygonGetTransactionsCount = polygonGetTransactionsCount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const polygonGetCurrentBlock = async () => tatum_1.get(`/v3/polygon/block/current`);
exports.polygonGetCurrentBlock = polygonGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonGetBlock" target="_blank">Tatum API documentation</a>
 */
const polygonGetBlock = async (hash) => tatum_1.get(`/v3/polygon/block/${hash}`);
exports.polygonGetBlock = polygonGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonGetBalance" target="_blank">Tatum API documentation</a>
 */
const polygonGetAccountBalance = async (address) => tatum_1.get(`/v3/polygon/account/balance/${address}`);
exports.polygonGetAccountBalance = polygonGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonGetTransaction" target="_blank">Tatum API documentation</a>
 */
const polygonGetTransaction = async (hash) => tatum_1.get(`/v3/polygon/transaction/${hash}`);
exports.polygonGetTransaction = polygonGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/PolygonEstimateGas" target="_blank">Tatum API documentation</a>
 */
const polygonEstimateGas = (body) => tatum_1.post('/v3/polygon/gas', body, model_1.EstimateGasEth);
exports.polygonEstimateGas = polygonEstimateGas;
