// @tatumio/tatum/dist/src/blockchain/qtum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.qtumBroadcast = exports.estimateQtumFeePerByte = exports.estimateQtumFee = exports.getQtumBlock = exports.getQtumCurrentBlock = exports.getQtumBalance = exports.getQtumTransactions = exports.getQtumTransaction = exports.getQtumUTXOs = void 0;
const tatum_1 = require("../connector/tatum");
const getQtumUTXOs = async (address) => tatum_1.get(`/v3/qtum/utxo/${address}`);
exports.getQtumUTXOs = getQtumUTXOs;
const getQtumTransaction = async (id) => tatum_1.get(`/v3/qtum/transaction/${id}`);
exports.getQtumTransaction = getQtumTransaction;
const getQtumTransactions = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/qtum/transactions/address/${address}/?pageSize=${pageSize}&offset=${offset}`);
exports.getQtumTransactions = getQtumTransactions;
const getQtumBalance = async (address) => tatum_1.get(`/v3/qtum/account/balance/${address}`);
exports.getQtumBalance = getQtumBalance;
const getQtumCurrentBlock = async () => tatum_1.get(`/v3/qtum/block/current`);
exports.getQtumCurrentBlock = getQtumCurrentBlock;
const getQtumBlock = async (hash) => tatum_1.get(`/v3/qtum/block/${hash}`);
exports.getQtumBlock = getQtumBlock;
const estimateQtumFee = async (nblocks) => tatum_1.get(`/v3/qtum/transactions/gas/${nblocks}`);
exports.estimateQtumFee = estimateQtumFee;
const estimateQtumFeePerByte = async (nblocks) => tatum_1.get(`/v3/qtum/transactions/gasbytes/${nblocks}`);
exports.estimateQtumFeePerByte = estimateQtumFeePerByte;
const qtumBroadcast = async (rawtx) => tatum_1.post('v3/qtum/broadcast', { rawtx });
exports.qtumBroadcast = qtumBroadcast;
