// @tatumio/tatum/dist/src/blockchain/quorum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.quorumUnlockAccount = exports.quorumCreateAccount = exports.quorumGetTransactionReceipt = exports.quorumGetTransaction = exports.quorumGetBlock = exports.quorumGetCurrentBlock = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const quorumGetCurrentBlock = async () => tatum_1.get(`/v3/quorum/block/current`);
exports.quorumGetCurrentBlock = quorumGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetBlock" target="_blank">Tatum API documentation</a>
 */
const quorumGetBlock = async (hash) => tatum_1.get(`/v3/quorum/block/${hash}`);
exports.quorumGetBlock = quorumGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetTransaction" target="_blank">Tatum API documentation</a>
 */
const quorumGetTransaction = async (hash) => tatum_1.get(`/v3/quorum/transaction/${hash}`);
exports.quorumGetTransaction = quorumGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetTransactionReceipt" target="_blank">Tatum API documentation</a>
 */
const quorumGetTransactionReceipt = async (hash) => tatum_1.get(`/v3/quorum/transaction/${hash}/receipt`);
exports.quorumGetTransactionReceipt = quorumGetTransactionReceipt;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGenerateAccount" target="_blank">Tatum API documentation</a>
 */
const quorumCreateAccount = async (body) => tatum_1.post(`/v3/quorum/account`, body, model_1.AccountPassword);
exports.quorumCreateAccount = quorumCreateAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumUnlockAccount" target="_blank">Tatum API documentation</a>
 */
const quorumUnlockAccount = async (address, body) => tatum_1.post(`/v3/quorum/account/${address}/unlock`, body, model_1.AccountPassword);
exports.quorumUnlockAccount = quorumUnlockAccount;
