// @tatumio/tatum/dist/src/blockchain/scrypta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scryptaGetTransaction = exports.scryptaGetUnspentForAccount = exports.scryptaGetTxForAccount = exports.scryptaGetUTXO = exports.scryptaGetBlockHash = exports.scryptaGetBlock = exports.scryptaGetCurrentBlock = exports.scryptaBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BroadcastsignedScryptatransaction" target="_blank">Tatum API documentation</a>
 */
const scryptaBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/scrypta/broadcast`, { txData, signatureId });
exports.scryptaBroadcast = scryptaBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
const scryptaGetCurrentBlock = async () => tatum_1.get(`/v3/scrypta/info`);
exports.scryptaGetCurrentBlock = scryptaGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlock" target="_blank">Tatum API documentation</a>
 */
const scryptaGetBlock = async (hash) => tatum_1.get(`/v3/scrypta/block/${hash}`);
exports.scryptaGetBlock = scryptaGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlockHash" target="_blank">Tatum API documentation</a>
 */
const scryptaGetBlockHash = async (i) => tatum_1.get(`/v3/scrypta/block/hash/${i}`);
exports.scryptaGetBlockHash = scryptaGetBlockHash;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetUTXO" target="_blank">Tatum API documentation</a>
 */
const scryptaGetUTXO = async (hash, i) => tatum_1.get(`/v3/scrypta/utxo/${hash}/${i}`);
exports.scryptaGetUTXO = scryptaGetUTXO;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
const scryptaGetTxForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/scrypta/transaction/address/${address}?pageSize=${pageSize}&offset=${offset}`);
exports.scryptaGetTxForAccount = scryptaGetTxForAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
const scryptaGetUnspentForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/scrypta/utxo/${address}?pageSize=${pageSize}&offset=${offset}`);
exports.scryptaGetUnspentForAccount = scryptaGetUnspentForAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
const scryptaGetTransaction = async (hash) => tatum_1.get(`/v3/scrypta/transaction/${hash}`);
exports.scryptaGetTransaction = scryptaGetTransaction;
