// @tatumio/tatum/dist/src/blockchain/tron.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tronGetAccount = exports.tronGetTransaction = exports.tronGetTrc10Detail = exports.tronGetBlock = exports.tronGetCurrentBlock = exports.tronBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronBroadcast" target="_blank">Tatum API documentation</a>
 */
const tronBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/tron/broadcast`, { txData, signatureId });
exports.tronBroadcast = tronBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const tronGetCurrentBlock = async () => tatum_1.get(`/v3/tron/info`);
exports.tronGetCurrentBlock = tronGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetBlock" target="_blank">Tatum API documentation</a>
 */
const tronGetBlock = async (hash) => tatum_1.get(`/v3/tron/block/${hash}`);
exports.tronGetBlock = tronGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronTrc10Detail" target="_blank">Tatum API documentation</a>
 */
const tronGetTrc10Detail = async (id) => tatum_1.get(`/v3/tron/trc10/detail/${id}`);
exports.tronGetTrc10Detail = tronGetTrc10Detail;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetTransaction" target="_blank">Tatum API documentation</a>
 */
const tronGetTransaction = async (hash) => tatum_1.get(`/v3/tron/transaction/${hash}`);
exports.tronGetTransaction = tronGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetAccount" target="_blank">Tatum API documentation</a>
 */
const tronGetAccount = async (address) => tatum_1.get(`/v3/tron/account/${address}`);
exports.tronGetAccount = tronGetAccount;
