// @tatumio/tatum/dist/src/blockchain/vet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.vetGetTransactionReceipt = exports.vetGetTransaction = exports.vetGetAccountEnergy = exports.vetGetAccountBalance = exports.vetGetBlock = exports.vetGetCurrentBlock = exports.vetEstimateGas = exports.vetBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetBroadcast" target="_blank">Tatum API documentation</a>
 */
const vetBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/vet/broadcast`, { txData, signatureId });
exports.vetBroadcast = vetBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetEstimateGas" target="_blank">Tatum API documentation</a>
 */
const vetEstimateGas = async (body) => tatum_1.post(`/v3/vet/broadcast/transaction/gas`, body, model_1.EstimateGasVet);
exports.vetEstimateGas = vetEstimateGas;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const vetGetCurrentBlock = async () => tatum_1.get(`/v3/vet/current`);
exports.vetGetCurrentBlock = vetGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetBlock" target="_blank">Tatum API documentation</a>
 */
const vetGetBlock = async (hash) => tatum_1.get(`/v3/vet/block/${hash}`);
exports.vetGetBlock = vetGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetBalance" target="_blank">Tatum API documentation</a>
 */
const vetGetAccountBalance = async (address) => tatum_1.get(`/v3/vet/account/balance/${address}`);
exports.vetGetAccountBalance = vetGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetEnergy" target="_blank">Tatum API documentation</a>
 */
const vetGetAccountEnergy = async (address) => tatum_1.get(`/v3/vet/account/energy/${address}`);
exports.vetGetAccountEnergy = vetGetAccountEnergy;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetTransaction" target="_blank">Tatum API documentation</a>
 */
const vetGetTransaction = async (hash) => tatum_1.get(`/v3/vet/transaction/${hash}`);
exports.vetGetTransaction = vetGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/VetGetTransactionReceipt" target="_blank">Tatum API documentation</a>
 */
const vetGetTransactionReceipt = async (hash) => tatum_1.get(`/v3/vet/transaction/${hash}/receipt`);
exports.vetGetTransactionReceipt = vetGetTransactionReceipt;
