// @tatumio/tatum/dist/src/blockchain/xdc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xdcEstimateGas = exports.xdcGetTransaction = exports.xdcGetAccountErc20Balance = exports.xdcGetAccountBalance = exports.xdcGetBlock = exports.xdcGetCurrentBlock = exports.xdcGetTransactionsCount = exports.xdcBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcBroadcast" target="_blank">Tatum API documentation</a>
 */
const xdcBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/xdc/broadcast`, { txData, signatureId });
exports.xdcBroadcast = xdcBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
const xdcGetTransactionsCount = async (address) => tatum_1.get(`/v3/xdc/transaction/count/${address}`);
exports.xdcGetTransactionsCount = xdcGetTransactionsCount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
const xdcGetCurrentBlock = async () => tatum_1.get(`/v3/xdc/block/current`);
exports.xdcGetCurrentBlock = xdcGetCurrentBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcGetBlock" target="_blank">Tatum API documentation</a>
 */
const xdcGetBlock = async (hash) => tatum_1.get(`/v3/xdc/block/${hash}`);
exports.xdcGetBlock = xdcGetBlock;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcGetBalance" target="_blank">Tatum API documentation</a>
 */
const xdcGetAccountBalance = async (address) => tatum_1.get(`/v3/xdc/account/balance/${address}`);
exports.xdcGetAccountBalance = xdcGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcErc20GetBalance" target="_blank">Tatum API documentation</a>
 */
const xdcGetAccountErc20Balance = async (address, contractAddress) => tatum_1.get(`/v3/xdc/account/balance/erc20/${address}?contractAddress=${contractAddress}`);
exports.xdcGetAccountErc20Balance = xdcGetAccountErc20Balance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcGetTransaction" target="_blank">Tatum API documentation</a>
 */
const xdcGetTransaction = async (hash) => tatum_1.get(`/v3/xdc/transaction/${hash}`);
exports.xdcGetTransaction = xdcGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XdcEstimateGas" target="_blank">Tatum API documentation</a>
 */
const xdcEstimateGas = (body) => tatum_1.post('/v3/xdc/gas', body, model_1.EstimateGasEth);
exports.xdcEstimateGas = xdcEstimateGas;
