// @tatumio/tatum/dist/src/blockchain/xlm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xlmGetAccountTransactions = exports.xlmGetTransaction = exports.xlmGetLedgerTx = exports.xlmGetLedger = exports.xlmGetFee = exports.xlmGetCurrentLedger = exports.xlmBroadcast = exports.xlmGetAccountInfo = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetAccountInfo" target="_blank">Tatum API documentation</a>
 */
const xlmGetAccountInfo = async (account) => tatum_1.get(`/v3/xlm/account/${account}`);
exports.xlmGetAccountInfo = xlmGetAccountInfo;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmBroadcast" target="_blank">Tatum API documentation</a>
 */
const xlmBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/xlm/broadcast`, { txData, signatureId });
exports.xlmBroadcast = xlmBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLastClosedLedger" target="_blank">Tatum API documentation</a>
 */
const xlmGetCurrentLedger = async () => tatum_1.get(`/v3/xlm/info`);
exports.xlmGetCurrentLedger = xlmGetCurrentLedger;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetFee" target="_blank">Tatum API documentation</a>
 */
const xlmGetFee = async () => tatum_1.get(`/v3/xlm/fee`);
exports.xlmGetFee = xlmGetFee;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLedger" target="_blank">Tatum API documentation</a>
 */
const xlmGetLedger = async (i) => tatum_1.get(`/v3/xlm/ledger/${i}`);
exports.xlmGetLedger = xlmGetLedger;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLedgerTx" target="_blank">Tatum API documentation</a>
 */
const xlmGetLedgerTx = async (i) => tatum_1.get(`/v3/xlm/ledger/${i}/transaction`);
exports.xlmGetLedgerTx = xlmGetLedgerTx;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetTransaction" target="_blank">Tatum API documentation</a>
 */
const xlmGetTransaction = async (hash) => tatum_1.get(`/v3/xlm/transaction/${hash}`);
exports.xlmGetTransaction = xlmGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetAccountTx" target="_blank">Tatum API documentation</a>
 */
const xlmGetAccountTransactions = async (address) => tatum_1.get(`/v3/xlm/account/tx/${address}`);
exports.xlmGetAccountTransactions = xlmGetAccountTransactions;
