// @tatumio/tatum/dist/src/blockchain/xrp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xrpGetAccountTransactions = exports.xrpGetTransaction = exports.xrpGetAccountBalance = exports.xrpGetLedger = exports.xrpGetCurrentLedger = exports.xrpBroadcast = exports.xrpGetAccountInfo = exports.xrpGetFee = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetFee" target="_blank">Tatum API documentation</a>
 */
const xrpGetFee = async () => tatum_1.get(`/v3/xrp/fee`);
exports.xrpGetFee = xrpGetFee;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountInfo" target="_blank">Tatum API documentation</a>
 */
const xrpGetAccountInfo = async (account) => tatum_1.get(`/v3/xrp/account/${account}`);
exports.xrpGetAccountInfo = xrpGetAccountInfo;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpBroadcast" target="_blank">Tatum API documentation</a>
 */
const xrpBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/xrp/broadcast`, { txData, signatureId });
exports.xrpBroadcast = xrpBroadcast;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetLastClosedLedger" target="_blank">Tatum API documentation</a>
 */
const xrpGetCurrentLedger = async () => tatum_1.get(`/v3/xrp/info`);
exports.xrpGetCurrentLedger = xrpGetCurrentLedger;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetLedger" target="_blank">Tatum API documentation</a>
 */
const xrpGetLedger = async (i) => tatum_1.get(`/v3/xrp/ledger/${i}`);
exports.xrpGetLedger = xrpGetLedger;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountBalance" target="_blank">Tatum API documentation</a>
 */
const xrpGetAccountBalance = async (address) => tatum_1.get(`/v3/xrp/account/${address}/balance`);
exports.xrpGetAccountBalance = xrpGetAccountBalance;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetTransaction" target="_blank">Tatum API documentation</a>
 */
const xrpGetTransaction = async (hash) => tatum_1.get(`/v3/xrp/transaction/${hash}`);
exports.xrpGetTransaction = xrpGetTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountTx" target="_blank">Tatum API documentation</a>
 */
const xrpGetAccountTransactions = async (address, min, marker) => tatum_1.get(`/v3/xrp/account/tx/${address}${min ? `?min=${min}${marker ? `&marker=${marker}` : ''}` : marker ? `?marker=${marker}` : ''}`);
exports.xrpGetAccountTransactions = xrpGetAccountTransactions;
