// thorify/dist/utils/index.js
'use strict';
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.leftPadToBytes32 = exports.validBytes32OrError = exports.validAddressOrError = exports.validNumberOrDefault = exports.isFunction = exports.isUndefined = exports.isNull = exports.isObject = exports.isArray = exports.toInteger = exports.newNonce = exports.isHex = exports.sanitizeHex = exports.toPrefixedHex = void 0;
__exportStar(require("./params"), exports);
__exportStar(require("./option-formatters"), exports);
exports.toPrefixedHex = function (hexStr) {
    if (hexStr.indexOf('0x') === 0) {
        return hexStr;
    }
    else {
        return '0x' + hexStr;
    }
};
exports.sanitizeHex = function (hexStr) {
    if (hexStr.indexOf('0x') === 0) {
        return hexStr.substr(2);
    }
    else {
        return hexStr;
    }
};
exports.isHex = function (hex) {
    return !!hex && ((typeof hex === 'string') && /^(-0x|0x)?[0-9a-f]+$/i.test(hex));
};
exports.newNonce = function () {
    return Math.floor((new Date().getTime() / 0xffff) * Math.random() * 0xffff);
};
exports.toInteger = function (input) {
    const num = Number.parseInt(input);
    if (Number.isInteger(num)) {
        return num;
    }
    else {
        return null;
    }
};
exports.isArray = function (o) {
    return Object.prototype.toString.call(o) === '[object Array]';
};
exports.isObject = function (o) {
    return Object.prototype.toString.call(o) === '[object Object]';
};
exports.isNull = function (o) {
    return Object.prototype.toString.call(o) === '[object Null]';
};
exports.isUndefined = function (o) {
    return Object.prototype.toString.call(o) === '[object Undefined]';
};
exports.isFunction = function (o) {
    return typeof o === 'function';
};
exports.validNumberOrDefault = function (value, defaultValue) {
    if (typeof value === 'number' && Number.isInteger(value)) {
        return Math.abs(value);
    }
    if (Number.isNaN(Number.parseInt(value)) === false) {
        return Math.abs(Number.parseInt(value));
    }
    return defaultValue;
};
exports.validAddressOrError = function (input, msg = 'Invalid address string') {
    if (/^(-0x|0x)?[0-9a-fA-F]{40}$/i.test(input)) {
        return exports.toPrefixedHex(input);
    }
    else {
        throw new Error(msg);
    }
};
exports.validBytes32OrError = function (input, msg = 'Invalid hex string') {
    if (/^(-0x|0x)?[0-9a-fA-F]{64}$/i.test(input)) {
        return exports.toPrefixedHex(input);
    }
    else {
        throw new Error(msg);
    }
};
exports.leftPadToBytes32 = function (input) {
    return ('0x' + '0'.repeat(64 - exports.sanitizeHex(input).length) + exports.sanitizeHex(input));
};
