// thorify/dist/utils/option-formatters.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatLogQuery = exports.formatOptions = exports.formatRange = exports.fromETHBlockNumberOrHash = exports.fromETHBlockNumber = void 0;
const utils = require("./");
const OBJhas = (obj, p) => {
    return obj.hasOwnProperty(p);
};
exports.fromETHBlockNumber = function (blockNumber) {
    if (typeof blockNumber === 'number') {
        return blockNumber;
    }
    else if (typeof blockNumber === 'string') {
        if (blockNumber === 'earliest') {
            return 0;
        }
        else if (blockNumber === 'latest' || blockNumber === 'pending') {
            return 'best';
        }
        else {
            const num = utils.toInteger(blockNumber);
            return num || num === 0 ? num : 'best';
        }
    }
    else {
        return 'best';
    }
};
exports.fromETHBlockNumberOrHash = function (blockRevision) {
    if (/^(-0x|0x)?[0-9a-fA-F]{64}$/i.test(blockRevision)) {
        return blockRevision;
    }
    else {
        return exports.fromETHBlockNumber(blockRevision);
    }
};
exports.formatRange = function (range) {
    const defaultUnit = 'block';
    const ret = {
        unit: range.unit || defaultUnit,
        from: 0,
        to: Number.MAX_SAFE_INTEGER
    };
    if (ret.unit !== 'block' && ret.unit !== 'time') {
        ret.unit = defaultUnit;
    }
    if (range.unit === 'block') {
        if (OBJhas(range, 'from')) {
            const t = exports.fromETHBlockNumber(range.from);
            if (t === 'best') {
                ret.from = Number.MAX_SAFE_INTEGER;
            }
            else {
                ret.from = t;
            }
        }
        if (OBJhas(range, 'to')) {
            const t = exports.fromETHBlockNumber(range.to);
            if (t === 'best') {
                ret.to = Number.MAX_SAFE_INTEGER;
            }
            else {
                ret.to = t;
            }
        }
        return ret;
    }
    if (OBJhas(range, 'from')) {
        const t = utils.toInteger(range.from);
        if (t !== null) {
            ret.from = t;
        }
    }
    if (OBJhas(range, 'to')) {
        const t = utils.toInteger(range.to);
        if (t !== null) {
            ret.to = t;
        }
    }
    return ret;
};
exports.formatOptions = function (options) {
    const ret = {
        offset: 0,
        limit: utils.params.defaultLogLimit,
    };
    if (OBJhas(options, 'offset')) {
        const temp = utils.toInteger(options.offset);
        if (temp) {
            ret.offset = temp;
        }
    }
    if (OBJhas(options, 'limit')) {
        const temp = utils.toInteger(options.limit);
        if (temp) {
            ret.limit = temp;
        }
    }
    if (ret.limit > utils.params.maxLogLimit) {
        ret.limit = utils.params.maxLogLimit;
    }
    return ret;
};
exports.formatLogQuery = function (params) {
    let address = '';
    let order = 'ASC';
    if (params.address) {
        address = params.address;
    }
    if (params.order && (params.order.toUpperCase() === 'ASC' || params.order.toUpperCase() === 'DESC')) {
        order = params.order.toUpperCase();
    }
    const body = {
        criteriaSet: [],
        order,
    };
    if (params.options) {
        body.options = exports.formatOptions(params.options);
    }
    // discard fromBlock and toBlock if range presents
    if (params.range) {
        body.range = exports.formatRange(params.range);
    }
    else {
        const range = {
            unit: 'block',
            from: params.fromBlock || 0,
            to: params.toBlock || Number.MAX_SAFE_INTEGER
        };
        body.range = exports.formatRange(range);
    }
    const topics = [];
    if (params.topics && params.topics.length) {
        for (let i = 0; i < params.topics.length; i++) {
            if (typeof params.topics[i] === 'string') {
                topics.push({
                    name: 'topic' + i,
                    array: [params.topics[i]],
                });
            }
            else if (utils.isArray(params.topics[i]) && params.topics[i].length) {
                topics.push({
                    name: 'topic' + i,
                    array: params.topics[i],
                });
            }
        }
    }
    const outputTopic = function (inputTopics, index, receiver, current) {
        if (index === inputTopics.length) {
            const o = {};
            if (address) {
                current.address = address;
            }
            Object.assign(o, current);
            receiver.push(o);
            return;
        }
        for (const item of inputTopics[index].array) {
            current[inputTopics[index].name] = item;
            outputTopic(inputTopics, index + 1, receiver, current);
        }
    };
    if (topics.length) {
        outputTopic(topics, 0, body.criteriaSet, {});
    }
    if (!body.criteriaSet.length && address) {
        body.criteriaSet.push({
            address,
        });
    }
    return body;
};
