// @tatumio/tatum/dist/src/model/request/CreateSubscription.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateSubscription = exports.SubscriptionAttrCompleteBlockchainTx = exports.SubscriptionAttrUrlAddress = exports.SubscriptionAttrUrl = exports.SubscriptionAttrIncomingBlockchainTx = exports.SubscriptionAttrTxHistoryReport = exports.SubscriptionAttrOffchainWithdrawal = exports.SubscriptionAttrAccountBalanceLimit = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const SubscriptionType_1 = require("../response/ledger/SubscriptionType");
const Currency_1 = require("./Currency");
class Subscription {
}
class SubscriptionAttrAccountBalanceLimit extends Subscription {
}
__decorate([
    class_validator_1.MaxLength(38),
    class_validator_1.IsNotEmpty(),
    class_validator_1.Matches(/^[+]?((\d+(\.\d*)?)|(\.\d+))$/),
    __metadata("design:type", String)
], SubscriptionAttrAccountBalanceLimit.prototype, "limit", void 0);
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.IsIn(['account', 'available']),
    __metadata("design:type", String)
], SubscriptionAttrAccountBalanceLimit.prototype, "typeOfBalance", void 0);
exports.SubscriptionAttrAccountBalanceLimit = SubscriptionAttrAccountBalanceLimit;
class SubscriptionAttrOffchainWithdrawal extends Subscription {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(1, 30),
    class_validator_1.Matches(/^BTC|BNB|LTC|BCH|ETH|USDT|LEO|LINK|UNI|FREE|MKR|USDC|BAT|TUSD|PAX|PAXG|PLTC|XCON|MMY|[a-zA-Z0-9_]+$/),
    __metadata("design:type", String)
], SubscriptionAttrOffchainWithdrawal.prototype, "currency", void 0);
exports.SubscriptionAttrOffchainWithdrawal = SubscriptionAttrOffchainWithdrawal;
class SubscriptionAttrTxHistoryReport extends Subscription {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.IsNumber(),
    class_validator_1.Min(1),
    class_validator_1.Max(720),
    __metadata("design:type", Number)
], SubscriptionAttrTxHistoryReport.prototype, "interval", void 0);
exports.SubscriptionAttrTxHistoryReport = SubscriptionAttrTxHistoryReport;
class SubscriptionAttrIncomingBlockchainTx extends Subscription {
}
__decorate([
    class_validator_1.Length(24, 24),
    class_validator_1.IsNotEmpty(),
    __metadata("design:type", String)
], SubscriptionAttrIncomingBlockchainTx.prototype, "id", void 0);
__decorate([
    class_validator_1.IsUrl(),
    class_validator_1.IsNotEmpty(),
    class_validator_1.MaxLength(500),
    __metadata("design:type", String)
], SubscriptionAttrIncomingBlockchainTx.prototype, "url", void 0);
exports.SubscriptionAttrIncomingBlockchainTx = SubscriptionAttrIncomingBlockchainTx;
class SubscriptionAttrUrl extends Subscription {
}
__decorate([
    class_validator_1.IsUrl(),
    class_validator_1.IsNotEmpty(),
    class_validator_1.MaxLength(500),
    __metadata("design:type", String)
], SubscriptionAttrUrl.prototype, "url", void 0);
exports.SubscriptionAttrUrl = SubscriptionAttrUrl;
class SubscriptionAttrUrlAddress extends Subscription {
}
__decorate([
    class_validator_1.IsUrl(),
    class_validator_1.IsNotEmpty(),
    class_validator_1.MaxLength(500),
    __metadata("design:type", String)
], SubscriptionAttrUrlAddress.prototype, "url", void 0);
__decorate([
    class_validator_1.IsIn([
        Currency_1.Currency.SOL,
        Currency_1.Currency.ETH,
        Currency_1.Currency.MATIC,
        Currency_1.Currency.CELO,
        Currency_1.Currency.BTC,
        Currency_1.Currency.LTC,
        Currency_1.Currency.BCH,
        Currency_1.Currency.DOGE,
    ]),
    class_validator_1.IsNotEmpty(),
    __metadata("design:type", String)
], SubscriptionAttrUrlAddress.prototype, "chain", void 0);
__decorate([
    class_validator_1.Length(13, 128),
    class_validator_1.IsNotEmpty(),
    __metadata("design:type", String)
], SubscriptionAttrUrlAddress.prototype, "address", void 0);
exports.SubscriptionAttrUrlAddress = SubscriptionAttrUrlAddress;
class SubscriptionAttrCompleteBlockchainTx extends Subscription {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(1, 30),
    class_validator_1.Matches(/^BTC|ADA|LTC|FLOW|FUSD|ONE|XDC|DOGE|RMD|BSC|CELO|CEUR|CUSD|BETH|GMC|GMC_BSC|BUSD|BUSD_BSC|CAKE|BBTC|BADA|WBNB|BDOT|BXRP|BLTC|BBCH|BCH|ETH|USDT|WBTC|LEO|LINK|UNI|FREE|MKR|USDC|BAT|TUSD|PAX|PAXG|PLTC|XCON|MMY|[a-zA-Z0-9_]+$/),
    __metadata("design:type", String)
], SubscriptionAttrCompleteBlockchainTx.prototype, "currency", void 0);
exports.SubscriptionAttrCompleteBlockchainTx = SubscriptionAttrCompleteBlockchainTx;
class CreateSubscription {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.IsIn(Object.values(SubscriptionType_1.SubscriptionType)),
    __metadata("design:type", String)
], CreateSubscription.prototype, "type", void 0);
__decorate([
    class_validator_1.IsNotEmptyObject(),
    class_validator_1.ValidateNested(),
    class_transformer_1.Type(() => Subscription, {
        discriminator: {
            property: '__type',
            subTypes: [
                { value: SubscriptionAttrAccountBalanceLimit, name: 'ACCOUNT_BALANCE_LIMIT' },
                { value: SubscriptionAttrUrl, name: 'KMS_COMPLETED_TX' },
                { value: SubscriptionAttrUrl, name: 'KMS_FAILED_TX' },
                { value: SubscriptionAttrUrl, name: 'CUSTOMER_TRADE_MATCH' },
                { value: SubscriptionAttrUrl, name: 'TRANSACTION_IN_THE_BLOCK' },
                { value: SubscriptionAttrUrlAddress, name: 'ADDRESS_TRANSACTION' },
                { value: SubscriptionAttrOffchainWithdrawal, name: 'OFFCHAIN_WITHDRAWAL' },
                { value: SubscriptionAttrTxHistoryReport, name: 'TRANSACTION_HISTORY_REPORT' },
                { value: SubscriptionAttrIncomingBlockchainTx, name: 'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION' },
                { value: SubscriptionAttrCompleteBlockchainTx, name: 'COMPLETE_BLOCKCHAIN_TRANSACTION' },
                { value: SubscriptionAttrIncomingBlockchainTx, name: 'ACCOUNT_PENDING_BLOCKCHAIN_TRANSACTION' },
            ],
        },
    }),
    __metadata("design:type", Object)
], CreateSubscription.prototype, "attr", void 0);
exports.CreateSubscription = CreateSubscription;
