// @tatumio/tatum/dist/src/model/request/Currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BSC_BASED_CURRENCIES = exports.MATIC_BASED_CURRENCIES = exports.ETH_BASED_CURRENCIES = exports.MATIC20_CURRENCIES = exports.BEP20_CURRENCIES = exports.ERC20_CURRENCIES = exports.Currency = void 0;
var Currency;
(function (Currency) {
    Currency["BTC"] = "BTC";
    Currency["LUNA"] = "LUNA";
    Currency["LUNA_USD"] = "LUNA_USD";
    Currency["LUNA_KRW"] = "LUNA_KRW";
    Currency["BCH"] = "BCH";
    Currency["LTC"] = "LTC";
    Currency["CELO"] = "CELO";
    Currency["KLAY"] = "KLAY";
    Currency["ONE"] = "ONE";
    Currency["CUSD"] = "CUSD";
    Currency["CEUR"] = "CEUR";
    Currency["ETH"] = "ETH";
    Currency["XRP"] = "XRP";
    Currency["XLM"] = "XLM";
    Currency["DOGE"] = "DOGE";
    Currency["VET"] = "VET";
    Currency["NEO"] = "NEO";
    Currency["BNB"] = "BNB";
    Currency["BSC"] = "BSC";
    Currency["CAKE"] = "CAKE";
    Currency["BUSD_BSC"] = "BUSD_BSC";
    Currency["B2U_BSC"] = "B2U_BSC";
    Currency["BETH"] = "BETH";
    Currency["GAMEE"] = "GAMEE";
    Currency["BBTC"] = "BBTC";
    Currency["BADA"] = "BADA";
    Currency["RMD"] = "RMD";
    Currency["WBNB"] = "WBNB";
    Currency["BDOT"] = "BDOT";
    Currency["BXRP"] = "BXRP";
    Currency["BLTC"] = "BLTC";
    Currency["BBCH"] = "BBCH";
    Currency["MATIC"] = "MATIC";
    Currency["USDC_MATIC"] = "USDC_MATIC";
    Currency["USDC_BSC"] = "USDC_BSC";
    Currency["USDC_SOL"] = "USDC_SOL";
    Currency["USDT"] = "USDT";
    Currency["GMC"] = "GMC";
    Currency["GMC_BSC"] = "GMC_BSC";
    Currency["FLOW"] = "FLOW";
    Currency["FUSD"] = "FUSD";
    Currency["USDT_TRON"] = "USDT_TRON";
    Currency["INRT_TRON"] = "INRT_TRON";
    Currency["TRON"] = "TRON";
    Currency["LEO"] = "LEO";
    Currency["LINK"] = "LINK";
    Currency["LISK"] = "LISK";
    Currency["WBTC"] = "WBTC";
    Currency["UNI"] = "UNI";
    Currency["SOL"] = "SOL";
    Currency["FREE"] = "FREE";
    Currency["MKR"] = "MKR";
    Currency["USDC"] = "USDC";
    Currency["MATIC_ETH"] = "MATIC_ETH";
    Currency["BAT"] = "BAT";
    Currency["TUSD"] = "TUSD";
    Currency["BUSD"] = "BUSD";
    Currency["PAX"] = "PAX";
    Currency["PLTC"] = "PLTC";
    Currency["XCON"] = "XCON";
    Currency["REVV"] = "REVV";
    Currency["SAND"] = "SAND";
    Currency["MMY"] = "MMY";
    Currency["PAXG"] = "PAXG";
    Currency["HAG"] = "HAG";
    Currency["ADA"] = "ADA";
    Currency["XDC"] = "XDC";
    Currency["LATOKEN"] = "LATOKEN";
    Currency["USDT_MATIC"] = "USDT_MATIC";
    Currency["QTUM"] = "QTUM";
    Currency["EGLD"] = "EGLD";
    Currency["ALGO"] = "ALGO";
    Currency["KCS"] = "KCS";
    Currency["COIIN"] = "COIIN";
    Currency["COIIN_BSC"] = "COIIN_BSC";
    Currency["GLMR"] = "GLMR";
    Currency["EOS"] = "EOS";
    Currency["AVAX"] = "AVAX";
})(Currency = exports.Currency || (exports.Currency = {}));
exports.ERC20_CURRENCIES = [
    Currency.USDT.toString(),
    Currency.LATOKEN.toString(),
    Currency.LEO.toString(),
    Currency.LINK.toString(),
    Currency.UNI.toString(),
    Currency.FREE.toString(),
    Currency.MKR.toString(),
    Currency.BUSD.toString(),
    Currency.USDC.toString(),
    Currency.MATIC_ETH.toString(),
    Currency.BAT.toString(),
    Currency.WBTC.toString(),
    Currency.TUSD.toString(),
    Currency.PAX.toString(),
    Currency.PAXG.toString(),
    Currency.PLTC.toString(),
    Currency.XCON.toString(),
    Currency.MMY.toString(),
    Currency.COIIN.toString(),
    Currency.REVV.toString(),
    Currency.SAND.toString(),
    Currency.GMC.toString(),
];
exports.BEP20_CURRENCIES = [
    Currency.BETH.toString(),
    Currency.BBTC.toString(),
    Currency.RMD.toString(),
    Currency.USDC_BSC.toString(),
    Currency.COIIN_BSC.toString(),
    Currency.B2U_BSC.toString(),
    Currency.BADA.toString(),
    Currency.WBNB.toString(),
    Currency.GMC_BSC.toString(),
    Currency.BDOT.toString(),
    Currency.BXRP.toString(),
    Currency.BLTC.toString(),
    Currency.BBCH.toString(),
    Currency.HAG.toString(),
    Currency.CAKE.toString(),
    Currency.BUSD_BSC.toString(),
];
exports.MATIC20_CURRENCIES = [
    Currency.USDC_MATIC.toString(),
    Currency.GAMEE.toString(),
    Currency.USDT_MATIC.toString(),
];
exports.ETH_BASED_CURRENCIES = [
    Currency.ETH.toString(),
    ...exports.ERC20_CURRENCIES,
];
exports.MATIC_BASED_CURRENCIES = [
    Currency.MATIC.toString(),
    ...exports.MATIC20_CURRENCIES,
];
exports.BSC_BASED_CURRENCIES = [
    Currency.BSC.toString(),
    ...exports.BEP20_CURRENCIES,
];
