// @tatumio/tatum/dist/src/model/request/TransferBtcBasedBlockchain.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBtcBasedBlockchain = exports.To = exports.FromUTXO = exports.FromAddress = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const SignatureIdValidator_1 = require("../validation/SignatureIdValidator");
const TransferBtcValidator_1 = require("../validation/TransferBtcValidator");
class PrivateKeyOrSignatureIdBtcBased {
}
__decorate([
    class_validator_1.ValidateIf(o => o.privateKey || !o.signatureId),
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(52, 256),
    __metadata("design:type", String)
], PrivateKeyOrSignatureIdBtcBased.prototype, "privateKey", void 0);
__decorate([
    class_validator_1.ValidateIf(o => o.signatureId || !o.privateKey),
    class_validator_1.Validate(SignatureIdValidator_1.SignatureIdValidator),
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(36, 36),
    class_validator_1.IsUUID('4'),
    __metadata("design:type", String)
], PrivateKeyOrSignatureIdBtcBased.prototype, "signatureId", void 0);
class FromAddress extends PrivateKeyOrSignatureIdBtcBased {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(30, 110),
    __metadata("design:type", String)
], FromAddress.prototype, "address", void 0);
exports.FromAddress = FromAddress;
class FromUTXO extends PrivateKeyOrSignatureIdBtcBased {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(64, 64),
    __metadata("design:type", String)
], FromUTXO.prototype, "txHash", void 0);
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Min(0),
    class_validator_1.Max(4294967295),
    __metadata("design:type", Number)
], FromUTXO.prototype, "index", void 0);
exports.FromUTXO = FromUTXO;
class To {
}
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Length(30, 110),
    __metadata("design:type", String)
], To.prototype, "address", void 0);
__decorate([
    class_validator_1.IsNotEmpty(),
    class_validator_1.Min(0),
    __metadata("design:type", Number)
], To.prototype, "value", void 0);
exports.To = To;
/**
 * Transfer BTC based blockchain.
 *
 * This class is used for transferring assets of all BTC based blockchains.
 */
class TransferBtcBasedBlockchain {
}
__decorate([
    class_validator_1.ValidateIf(o => (o.fromUTXO && o.fromAddress) || !o.fromUTXO),
    class_validator_1.Validate(TransferBtcValidator_1.TransferBtcValidator),
    class_validator_1.IsNotEmpty(),
    class_validator_1.ValidateNested({ each: true }),
    class_transformer_1.Type(() => FromAddress),
    __metadata("design:type", Array)
], TransferBtcBasedBlockchain.prototype, "fromAddress", void 0);
__decorate([
    class_validator_1.ValidateIf(o => (o.fromUTXO && o.fromAddress) || !o.fromAddress),
    class_validator_1.Validate(TransferBtcValidator_1.TransferBtcValidator),
    class_validator_1.IsNotEmpty(),
    class_validator_1.ValidateNested({ each: true }),
    class_transformer_1.Type(() => FromUTXO),
    __metadata("design:type", Array)
], TransferBtcBasedBlockchain.prototype, "fromUTXO", void 0);
__decorate([
    class_validator_1.ArrayNotEmpty(),
    class_validator_1.ValidateNested({ each: true }),
    class_transformer_1.Type(() => To),
    __metadata("design:type", Array)
], TransferBtcBasedBlockchain.prototype, "to", void 0);
exports.TransferBtcBasedBlockchain = TransferBtcBasedBlockchain;
