// @tatumio/tatum/dist/src/model/response/ledger/SubscriptionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionType = void 0;
/**
 * SubscriptionType
 */
var SubscriptionType;
(function (SubscriptionType) {
    SubscriptionType["OFFCHAIN_WITHDRAWAL"] = "OFFCHAIN_WITHDRAWAL";
    SubscriptionType["ACCOUNT_BALANCE_LIMIT"] = "ACCOUNT_BALANCE_LIMIT";
    SubscriptionType["ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION"] = "ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION";
    SubscriptionType["ACCOUNT_PENDING_BLOCKCHAIN_TRANSACTION"] = "ACCOUNT_PENDING_BLOCKCHAIN_TRANSACTION";
    SubscriptionType["KMS_FAILED_TX"] = "KMS_FAILED_TX";
    SubscriptionType["KMS_COMPLETED_TX"] = "KMS_COMPLETED_TX";
    SubscriptionType["TRANSACTION_HISTORY_REPORT"] = "TRANSACTION_HISTORY_REPORT";
    SubscriptionType["CUSTOMER_TRADE_MATCH"] = "CUSTOMER_TRADE_MATCH";
    SubscriptionType["CUSTOMER_PARTIAL_TRADE_MATCH"] = "CUSTOMER_PARTIAL_TRADE_MATCH";
    SubscriptionType["TRANSACTION_IN_THE_BLOCK"] = "TRANSACTION_IN_THE_BLOCK";
    SubscriptionType["ADDRESS_TRANSACTION"] = "ADDRESS_TRANSACTION";
})(SubscriptionType = exports.SubscriptionType || (exports.SubscriptionType = {}));
