// @tatumio/tatum/dist/src/model/response/ledger/VC.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePairEnum = void 0;
var BasePairEnum;
(function (BasePairEnum) {
    BasePairEnum["AED"] = "AED";
    BasePairEnum["AFN"] = "AFN";
    BasePairEnum["ALL"] = "ALL";
    BasePairEnum["AMD"] = "AMD";
    BasePairEnum["ANG"] = "ANG";
    BasePairEnum["AOA"] = "AOA";
    BasePairEnum["ARS"] = "ARS";
    BasePairEnum["AUD"] = "AUD";
    BasePairEnum["AWG"] = "AWG";
    BasePairEnum["AZN"] = "AZN";
    BasePairEnum["BAM"] = "BAM";
    BasePairEnum["BAT"] = "BAT";
    BasePairEnum["BBD"] = "BBD";
    BasePairEnum["BCH"] = "BCH";
    BasePairEnum["BDT"] = "BDT";
    BasePairEnum["BGN"] = "BGN";
    BasePairEnum["BHD"] = "BHD";
    BasePairEnum["BIF"] = "BIF";
    BasePairEnum["BMD"] = "BMD";
    BasePairEnum["BND"] = "BND";
    BasePairEnum["BOB"] = "BOB";
    BasePairEnum["BRL"] = "BRL";
    BasePairEnum["BSD"] = "BSD";
    BasePairEnum["BTC"] = "BTC";
    BasePairEnum["BTN"] = "BTN";
    BasePairEnum["BWP"] = "BWP";
    BasePairEnum["BYN"] = "BYN";
    BasePairEnum["BYR"] = "BYR";
    BasePairEnum["BZD"] = "BZD";
    BasePairEnum["CAD"] = "CAD";
    BasePairEnum["CDF"] = "CDF";
    BasePairEnum["CHF"] = "CHF";
    BasePairEnum["CLF"] = "CLF";
    BasePairEnum["CLP"] = "CLP";
    BasePairEnum["CNY"] = "CNY";
    BasePairEnum["COP"] = "COP";
    BasePairEnum["CRC"] = "CRC";
    BasePairEnum["CUC"] = "CUC";
    BasePairEnum["CUP"] = "CUP";
    BasePairEnum["CVE"] = "CVE";
    BasePairEnum["CZK"] = "CZK";
    BasePairEnum["DJF"] = "DJF";
    BasePairEnum["DKK"] = "DKK";
    BasePairEnum["DOP"] = "DOP";
    BasePairEnum["DZD"] = "DZD";
    BasePairEnum["EGP"] = "EGP";
    BasePairEnum["ERN"] = "ERN";
    BasePairEnum["ETB"] = "ETB";
    BasePairEnum["ETH"] = "ETH";
    BasePairEnum["EUR"] = "EUR";
    BasePairEnum["FJD"] = "FJD";
    BasePairEnum["FKP"] = "FKP";
    BasePairEnum["FREE"] = "FREE";
    BasePairEnum["GBP"] = "GBP";
    BasePairEnum["GEL"] = "GEL";
    BasePairEnum["GGP"] = "GGP";
    BasePairEnum["GHS"] = "GHS";
    BasePairEnum["GIP"] = "GIP";
    BasePairEnum["GMD"] = "GMD";
    BasePairEnum["GNF"] = "GNF";
    BasePairEnum["GTQ"] = "GTQ";
    BasePairEnum["GYD"] = "GYD";
    BasePairEnum["HKD"] = "HKD";
    BasePairEnum["HNL"] = "HNL";
    BasePairEnum["HRK"] = "HRK";
    BasePairEnum["HTG"] = "HTG";
    BasePairEnum["HUF"] = "HUF";
    BasePairEnum["IDR"] = "IDR";
    BasePairEnum["ILS"] = "ILS";
    BasePairEnum["IMP"] = "IMP";
    BasePairEnum["INR"] = "INR";
    BasePairEnum["IQD"] = "IQD";
    BasePairEnum["IRR"] = "IRR";
    BasePairEnum["ISK"] = "ISK";
    BasePairEnum["JEP"] = "JEP";
    BasePairEnum["JMD"] = "JMD";
    BasePairEnum["JOD"] = "JOD";
    BasePairEnum["JPY"] = "JPY";
    BasePairEnum["KES"] = "KES";
    BasePairEnum["KGS"] = "KGS";
    BasePairEnum["KHR"] = "KHR";
    BasePairEnum["KMF"] = "KMF";
    BasePairEnum["KPW"] = "KPW";
    BasePairEnum["KRW"] = "KRW";
    BasePairEnum["KWD"] = "KWD";
    BasePairEnum["KYD"] = "KYD";
    BasePairEnum["KZT"] = "KZT";
    BasePairEnum["LAK"] = "LAK";
    BasePairEnum["LBP"] = "LBP";
    BasePairEnum["LEO"] = "LEO";
    BasePairEnum["LINK"] = "LINK";
    BasePairEnum["UNI"] = "UNI";
    BasePairEnum["LKR"] = "LKR";
    BasePairEnum["LRD"] = "LRD";
    BasePairEnum["LSL"] = "LSL";
    BasePairEnum["LTC"] = "LTC";
    BasePairEnum["LTL"] = "LTL";
    BasePairEnum["LVL"] = "LVL";
    BasePairEnum["LYD"] = "LYD";
    BasePairEnum["MAD"] = "MAD";
    BasePairEnum["MDL"] = "MDL";
    BasePairEnum["MGA"] = "MGA";
    BasePairEnum["MKD"] = "MKD";
    BasePairEnum["MKR"] = "MKR";
    BasePairEnum["MMK"] = "MMK";
    BasePairEnum["MMY"] = "MMY";
    BasePairEnum["MNT"] = "MNT";
    BasePairEnum["MOP"] = "MOP";
    BasePairEnum["MRO"] = "MRO";
    BasePairEnum["MUR"] = "MUR";
    BasePairEnum["MVR"] = "MVR";
    BasePairEnum["MWK"] = "MWK";
    BasePairEnum["MXN"] = "MXN";
    BasePairEnum["MYR"] = "MYR";
    BasePairEnum["MZN"] = "MZN";
    BasePairEnum["NAD"] = "NAD";
    BasePairEnum["NGN"] = "NGN";
    BasePairEnum["NIO"] = "NIO";
    BasePairEnum["NOK"] = "NOK";
    BasePairEnum["NPR"] = "NPR";
    BasePairEnum["NZD"] = "NZD";
    BasePairEnum["OMR"] = "OMR";
    BasePairEnum["PAB"] = "PAB";
    BasePairEnum["PAX"] = "PAX";
    BasePairEnum["PAXG"] = "PAXG";
    BasePairEnum["PEN"] = "PEN";
    BasePairEnum["PGK"] = "PGK";
    BasePairEnum["PHP"] = "PHP";
    BasePairEnum["PKR"] = "PKR";
    BasePairEnum["PLN"] = "PLN";
    BasePairEnum["PLTC"] = "PLTC";
    BasePairEnum["PYG"] = "PYG";
    BasePairEnum["QAR"] = "QAR";
    BasePairEnum["RON"] = "RON";
    BasePairEnum["RSD"] = "RSD";
    BasePairEnum["RUB"] = "RUB";
    BasePairEnum["RWF"] = "RWF";
    BasePairEnum["SAR"] = "SAR";
    BasePairEnum["SBD"] = "SBD";
    BasePairEnum["SCR"] = "SCR";
    BasePairEnum["SDG"] = "SDG";
    BasePairEnum["SEK"] = "SEK";
    BasePairEnum["SGD"] = "SGD";
    BasePairEnum["SHP"] = "SHP";
    BasePairEnum["SLL"] = "SLL";
    BasePairEnum["SOS"] = "SOS";
    BasePairEnum["SRD"] = "SRD";
    BasePairEnum["STD"] = "STD";
    BasePairEnum["SVC"] = "SVC";
    BasePairEnum["SYP"] = "SYP";
    BasePairEnum["SZL"] = "SZL";
    BasePairEnum["THB"] = "THB";
    BasePairEnum["TJS"] = "TJS";
    BasePairEnum["TMT"] = "TMT";
    BasePairEnum["TND"] = "TND";
    BasePairEnum["TOP"] = "TOP";
    BasePairEnum["TRY"] = "TRY";
    BasePairEnum["TTD"] = "TTD";
    BasePairEnum["TUSD"] = "TUSD";
    BasePairEnum["TWD"] = "TWD";
    BasePairEnum["TZS"] = "TZS";
    BasePairEnum["UAH"] = "UAH";
    BasePairEnum["UGX"] = "UGX";
    BasePairEnum["USD"] = "USD";
    BasePairEnum["USDC"] = "USDC";
    BasePairEnum["USDT"] = "USDT";
    BasePairEnum["UYU"] = "UYU";
    BasePairEnum["UZS"] = "UZS";
    BasePairEnum["VEF"] = "VEF";
    BasePairEnum["VND"] = "VND";
    BasePairEnum["VUV"] = "VUV";
    BasePairEnum["WST"] = "WST";
    BasePairEnum["XAF"] = "XAF";
    BasePairEnum["XAG"] = "XAG";
    BasePairEnum["XAU"] = "XAU";
    BasePairEnum["XCD"] = "XCD";
    BasePairEnum["XCON"] = "XCON";
    BasePairEnum["XDR"] = "XDR";
    BasePairEnum["XLM"] = "XLM";
    BasePairEnum["XOF"] = "XOF";
    BasePairEnum["XPF"] = "XPF";
    BasePairEnum["XRP"] = "XRP";
    BasePairEnum["YER"] = "YER";
    BasePairEnum["ZAR"] = "ZAR";
    BasePairEnum["ZMK"] = "ZMK";
    BasePairEnum["ZMW"] = "ZMW";
    BasePairEnum["ZWL"] = "ZWL";
})(BasePairEnum = exports.BasePairEnum || (exports.BasePairEnum = {}));
