// @tatumio/tatum/dist/src/ledger/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountBalance = exports.getAllAccounts = exports.getAccountsByCustomerId = exports.unfreezeAccount = exports.freezeAccount = exports.deactivateAccount = exports.activateAccount = exports.deleteBlockedAmountForAccount = exports.deleteBlockedAmountWithTransaction = exports.deleteBlockedAmount = exports.blockAmount = exports.getBlockedAmountsByAccountId = exports.createAccounts = exports.updateAccount = exports.generateAccount = exports.createAccount = exports.getAccountById = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const BlockageTransaction_1 = require("../model/request/BlockageTransaction");
const CreateAccountsBatch_1 = require("../model/request/CreateAccountsBatch");
const UpdateAccount_1 = require("../model/request/UpdateAccount");
const SubscriptionType_1 = require("../model/response/ledger/SubscriptionType");
const offchain_1 = require("../offchain");
const wallet_1 = require("../wallet");
const subscription_1 = require("./subscription");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountByAccountId" target="_blank">Tatum API documentation</a>
 */
const getAccountById = async (id) => tatum_1.get(`/v3/ledger/account/${id}`);
exports.getAccountById = getAccountById;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createAccount" target="_blank">Tatum API documentation</a>
 */
const createAccount = async (account) => tatum_1.post('/v3/ledger/account', account, model_1.CreateAccount);
exports.createAccount = createAccount;
/**
 * Abstraction unification endpoint for creating new ledger account, optionally added wallet generation, generating deposit blockchain address
 * and register incoming TX webhook notification.
 * @param account Account to be created.
 * @param generateNewWallet If new wallet should be created as well
 * @param testnet if we are using testnet or not
 * @param webhookUrl optional URL, where webhook will be post for every incoming blockchain transaction to the address
 */
const generateAccount = async (account, generateNewWallet = true, testnet = true, webhookUrl) => {
    let w;
    if (generateNewWallet) {
        // @ts-ignore
        w = await wallet_1.generateWallet(account.currency, testnet);
        // @ts-ignore
        account.xpub = w.xpub || w.address;
    }
    const a = await exports.createAccount(account);
    const address = await offchain_1.generateDepositAddress(a.id);
    if (webhookUrl) {
        await subscription_1.createNewSubscription({ type: SubscriptionType_1.SubscriptionType.ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION, attr: { url: webhookUrl, id: a.id } });
    }
    return { account: a, address, wallet: w };
};
exports.generateAccount = generateAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/updateAccountByAccountId" target="_blank">Tatum API documentation</a>
 */
const updateAccount = async (id, account) => tatum_1.put(`/v3/ledger/account/${id}`, account, UpdateAccount_1.UpdateAccount);
exports.updateAccount = updateAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createAccountBatch" target="_blank">Tatum API documentation</a>
 */
const createAccounts = async (accounts) => tatum_1.post('/v3/ledger/account/batch', accounts, CreateAccountsBatch_1.CreateAccountsBatch);
exports.createAccounts = createAccounts;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getBlockAmount" target="_blank">Tatum API documentation</a>
 */
const getBlockedAmountsByAccountId = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account/block/${id}?pageSize=${pageSize}&offset=${offset}`);
exports.getBlockedAmountsByAccountId = getBlockedAmountsByAccountId;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/blockAmount" target="_blank">Tatum API documentation</a>
 */
const blockAmount = async (id, block) => tatum_1.post(`/v3/ledger/account/block/${id}`, block, model_1.BlockAmount);
exports.blockAmount = blockAmount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteBlockAmount" target="_blank">Tatum API documentation</a>
 */
const deleteBlockedAmount = async (id) => tatum_1.httpDelete(`/v3/ledger/account/block/${id}`);
exports.deleteBlockedAmount = deleteBlockedAmount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/unblockAmountWithTransaction" target="_blank">Tatum API documentation</a>
 */
const deleteBlockedAmountWithTransaction = async (id, txData) => tatum_1.put(`/v3/ledger/account/block/${id}`, txData, BlockageTransaction_1.BlockageTransaction);
exports.deleteBlockedAmountWithTransaction = deleteBlockedAmountWithTransaction;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteAllBlockAmount" target="_blank">Tatum API documentation</a>
 */
const deleteBlockedAmountForAccount = async (id) => tatum_1.httpDelete(`/v3/ledger/account/block/account/${id}`);
exports.deleteBlockedAmountForAccount = deleteBlockedAmountForAccount;
/**
 * Activates deactivated account. If active account is deactivated nothing will happen.
 *
 * For more details, see <a href="https://tatum.io/apidoc#operation/activateAccount" target="_blank">Tatum API documentation</a>
 */
const activateAccount = async (id) => tatum_1.put(`v3/ledger/account/${id}/activate`);
exports.activateAccount = activateAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deactivateAccount" target="_blank">Tatum API documentation</a>
 */
const deactivateAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/deactivate`);
exports.deactivateAccount = deactivateAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/freezeAccount" target="_blank">Tatum API documentation</a>
 */
const freezeAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/freeze`);
exports.freezeAccount = freezeAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/unfreezeAccount" target="_blank">Tatum API documentation</a>
 */
const unfreezeAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/unfreeze`);
exports.unfreezeAccount = unfreezeAccount;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountsByCustomerId" target="_blank">Tatum API documentation</a>
 */
const getAccountsByCustomerId = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account/customer/${id}?pageSize=${pageSize}&offset=${offset}`);
exports.getAccountsByCustomerId = getAccountsByCustomerId;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAllAccounts" target="_blank">Tatum API documentation</a>
 */
const getAllAccounts = async (pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account?pageSize=${pageSize}&offset=${offset}`);
exports.getAllAccounts = getAllAccounts;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountBalance" target="_blank">Tatum API documentation</a>
 */
const getAccountBalance = async (id) => tatum_1.get(`/v3/ledger/account/${id}/balance`);
exports.getAccountBalance = getAccountBalance;
