// @tatumio/tatum/dist/src/ledger/customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disableCustomer = exports.enableCustomer = exports.deactivateCustomer = exports.activateCustomer = exports.updateCustomer = exports.getAllCustomers = exports.getCustomer = void 0;
const tatum_1 = require("../connector/tatum");
const CustomerUpdate_1 = require("../model/request/CustomerUpdate");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getCustomerByExternalId" target="_blank">Tatum API documentation</a>
 */
const getCustomer = async (id) => tatum_1.get(`/v3/ledger/customer/${id}`);
exports.getCustomer = getCustomer;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/findAllCustomers" target="_blank">Tatum API documentation</a>
 */
const getAllCustomers = async (pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/customer?pageSize=${pageSize}&offset=${offset}`);
exports.getAllCustomers = getAllCustomers;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/updateCustomer" target="_blank">Tatum API documentation</a>
 */
const updateCustomer = async (id, data) => tatum_1.put(`/v3/ledger/customer/${id}`, data, CustomerUpdate_1.CustomerUpdate);
exports.updateCustomer = updateCustomer;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/activateAccount" target="_blank">Tatum API documentation</a>
 */
const activateCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/activate`);
exports.activateCustomer = activateCustomer;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deactivateCustomer" target="_blank">Tatum API documentation</a>
 */
const deactivateCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/deactivate`);
exports.deactivateCustomer = deactivateCustomer;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/enableCustomer" target="_blank">Tatum API documentation</a>
 */
const enableCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/enable`);
exports.enableCustomer = enableCustomer;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/disableCustomer" target="_blank">Tatum API documentation</a>
 */
const disableCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/disable`);
exports.disableCustomer = disableCustomer;
