// @tatumio/tatum/dist/src/ledger/orderBook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteAccountTrades = exports.deleteTrade = exports.getTradeById = exports.storeTrade = exports.getActiveSellTrades = exports.getActiveBuyTrades = exports.getHistoricalTrades = void 0;
const tatum_1 = require("../connector/tatum");
const OrderBook_1 = require("../model/request/OrderBook");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getHistoricalTrades" target="_blank">Tatum API documentation</a>
 */
const getHistoricalTrades = async (pageSize = 50, offset = 0, id, pair) => {
    let url = `/v3/trade/history?pageSize=${pageSize}&offset=${offset}`;
    if (id) {
        url += `&id=${id}`;
    }
    if (pair) {
        url += `&pair=${pair}`;
    }
    return tatum_1.get(url);
};
exports.getHistoricalTrades = getHistoricalTrades;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getBuyTrades" target="_blank">Tatum API documentation</a>
 */
const getActiveBuyTrades = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/trade/buy?id=${id}&pageSize=${pageSize}&offset=${offset}`);
exports.getActiveBuyTrades = getActiveBuyTrades;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getSellTrades" target="_blank">Tatum API documentation</a>
 */
const getActiveSellTrades = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/trade/sell?id=${id}&pageSize=${pageSize}&offset=${offset}`);
exports.getActiveSellTrades = getActiveSellTrades;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/storeTrade" target="_blank">Tatum API documentation</a>
 */
const storeTrade = async (data) => tatum_1.post(`/v3/trade`, data, OrderBook_1.OrderBookRequest);
exports.storeTrade = storeTrade;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTradeById" target="_blank">Tatum API documentation</a>
 */
const getTradeById = async (id) => tatum_1.get(`/v3/trade/${id}`);
exports.getTradeById = getTradeById;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteTrade" target="_blank">Tatum API documentation</a>
 */
const deleteTrade = async (id) => tatum_1.httpDelete(`/v3/trade/${id}`);
exports.deleteTrade = deleteTrade;
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteAccountTrades" target="_blank">Tatum API documentation</a>
 */
const deleteAccountTrades = async (id) => tatum_1.httpDelete(`/v3/trade/account/${id}`);
exports.deleteAccountTrades = deleteAccountTrades;
