// @tatumio/tatum/dist/src/contracts/custodial/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustodialFullTokenWalletWithBatch = exports.CustodialFullTokenWallet = exports.Custodial_1155_TokenWalletWithBatch = exports.Custodial_1155_TokenWallet = exports.Custodial_721_TokenWalletWithBatch = exports.Custodial_721_TokenWallet = exports.Custodial_721_1155_TokenWalletWithBatch = exports.Custodial_721_1155_TokenWallet = exports.Custodial_20_TokenWalletWithBatch = exports.Custodial_20_TokenWallet = exports.Custodial_20_1155_TokenWalletWithBatch = exports.Custodial_20_1155_TokenWallet = exports.Custodial_20_721_TokenWalletWithBatch = exports.Custodial_20_721_TokenWallet = void 0;
exports.Custodial_20_721_TokenWallet = __importStar(require("./Custodial_20_721_TokenWallet"));
exports.Custodial_20_721_TokenWalletWithBatch = __importStar(require("./Custodial_20_721_TokenWalletWithBatch"));
exports.Custodial_20_1155_TokenWallet = __importStar(require("./Custodial_20_1155_TokenWallet"));
exports.Custodial_20_1155_TokenWalletWithBatch = __importStar(require("./Custodial_20_1155_TokenWalletWithBatch"));
exports.Custodial_20_TokenWallet = __importStar(require("./Custodial_20_TokenWallet"));
exports.Custodial_20_TokenWalletWithBatch = __importStar(require("./Custodial_20_TokenWalletWithBatch"));
exports.Custodial_721_1155_TokenWallet = __importStar(require("./Custodial_721_1155_TokenWallet"));
exports.Custodial_721_1155_TokenWalletWithBatch = __importStar(require("./Custodial_721_1155_TokenWalletWithBatch"));
exports.Custodial_721_TokenWallet = __importStar(require("./Custodial_721_TokenWallet"));
exports.Custodial_721_TokenWalletWithBatch = __importStar(require("./Custodial_721_TokenWalletWithBatch"));
exports.Custodial_1155_TokenWallet = __importStar(require("./Custodial_1155_TokenWallet"));
exports.Custodial_1155_TokenWalletWithBatch = __importStar(require("./Custodial_1155_TokenWalletWithBatch"));
exports.CustodialFullTokenWallet = __importStar(require("./CustodialFullTokenWallet"));
exports.CustodialFullTokenWalletWithBatch = __importStar(require("./CustodialFullTokenWalletWithBatch"));
