// @tatumio/tatum/dist/src/model/request/egld/EsdtToken.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EsdtToken = exports.EgldServiceType = void 0;
const class_validator_1 = require("class-validator");
var EgldServiceType;
(function (EgldServiceType) {
    EgldServiceType["issue"] = "issue";
    EgldServiceType["issueNonFungible"] = "issueNonFungible";
    EgldServiceType["issueSemiFungible"] = "issueSemiFungible";
    EgldServiceType["ESDTNFTCreate"] = "ESDTNFTCreate";
    EgldServiceType["stopNFTCreate"] = "stopNFTCreate";
    EgldServiceType["ESDTTransfer"] = "ESDTTransfer";
    EgldServiceType["mint"] = "mint";
    EgldServiceType["ESDTBurn"] = "ESDTBurn";
    EgldServiceType["pause"] = "pause";
    EgldServiceType["unPause"] = "unPause";
    EgldServiceType["freeze"] = "freeze";
    EgldServiceType["unFreeze"] = "unFreeze";
    EgldServiceType["wipe"] = "wipe";
    EgldServiceType["setSpecialRole"] = "setSpecialRole";
    EgldServiceType["unSetSpecialRole"] = "unSetSpecialRole";
    EgldServiceType["transferOwnership"] = "transferOwnership";
    EgldServiceType["controlChanges"] = "controlChanges";
    EgldServiceType["transferNFTCreateRole"] = "transferNFTCreateRole";
    EgldServiceType["ESDTNFTAddQuantity"] = "ESDTNFTAddQuantity";
    EgldServiceType["ESDTNFTBurn"] = "ESDTNFTBurn";
    EgldServiceType["freezeSingleNFT"] = "freezeSingleNFT";
    EgldServiceType["wipeSingleNFT"] = "wipeSingleNFT";
    EgldServiceType["unFreezeSingleNFT"] = "unFreezeSingleNFT";
    EgldServiceType["ESDTNFTTransfer"] = "ESDTNFTTransfer";
})(EgldServiceType = exports.EgldServiceType || (exports.EgldServiceType = {}));
class EsdtToken {
}
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.Length(1, 63),
    class_validator_1.IsIn(Object.keys(EgldServiceType)),
    __metadata("design:type", String)
], EsdtToken.prototype, "service", void 0);
__decorate([
    class_validator_1.ValidateIf(o => !['issue', 'issueNonFungible', 'issueSemiFungible'].includes(o.service)),
    class_validator_1.IsOptional(),
    class_validator_1.IsUppercase(),
    class_validator_1.Length(10, 17),
    __metadata("design:type", String)
], EsdtToken.prototype, "tokenId", void 0);
exports.EsdtToken = EsdtToken;
