// thorify/dist/provider/json-rpc.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONRPC = void 0;
const utils = require("../utils");
const thorifyResult = function (result) {
    // tricks for compatible with original web3 instance
    // non-objects or non-arrays doesn't need isThorified property since thorify just overwritten 3 formatters
    // which all accept object as input
    if (utils.isArray(result)) {
        result = result.map((item) => {
            Object.defineProperty(item, 'isThorified', { get: () => true });
            return item;
        });
    }
    else if (utils.isObject(result)) {
        Object.defineProperty(result, 'isThorified', { get: () => true });
    }
    return result;
};
class JSONRPC {
    constructor(payload) {
        this.id = payload.id;
        this.method = payload.method;
        this.params = payload.params;
    }
    makeResult(result) {
        return {
            id: this.id,
            jsonrpc: '2.0',
            result: thorifyResult(result),
        };
    }
    makeError(message) {
        return {
            id: this.id,
            jsonrpc: '2.0',
            error: {
                message,
            },
        };
    }
    makeSubResult(result) {
        return {
            id: this.id,
            jsonrpc: '2.0',
            method: this.method,
            params: {
                result: {
                    data: thorifyResult(result),
                },
                subscription: this.id,
            },
        };
    }
    makeSubError(error) {
        return {
            id: this.id,
            jsonrpc: '2.0',
            method: this.method,
            params: {
                result: {
                    error,
                },
                subscription: this.id,
            },
        };
    }
}
exports.JSONRPC = JSONRPC;
