// thorify/dist/provider/simple-http.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTP = exports.Method = void 0;
const XHR2 = require('xhr2');
var Method;
(function (Method) {
    Method[Method["GET"] = 0] = "GET";
    Method[Method["POST"] = 1] = "POST";
})(Method = exports.Method || (exports.Method = {}));
const post = function (url, body, timeout = 0) {
    return request(Method.POST, url, body, timeout);
};
const get = function (url, timeout = 0) {
    return request(Method.GET, url, null, timeout);
};
exports.HTTP = { get, post };
const request = function (method, url, body, timeout) {
    return new Promise((resolve, reject) => {
        const xhr = new XHR2();
        xhr.timeout = timeout;
        xhr.open(Method[method], url);
        if (method === Method.POST) {
            xhr.setRequestHeader('Content-Type', 'application/json');
        }
        xhr.onreadystatechange = () => {
            if (xhr.readyState === 4) {
                const res = {
                    Code: xhr.status,
                    Body: null,
                };
                if (xhr.status === 200) {
                    try {
                        res.Body = JSON.parse(xhr.responseText);
                    }
                    catch (e) {
                        return reject(new Error(`[thor-provider]Error parsing the response: ${e.message}`));
                    }
                }
                else if (xhr.responseText && xhr.responseText.length) {
                    res.Body = xhr.responseText;
                }
                return resolve(res);
            }
        };
        xhr.ontimeout = () => {
            return reject(new Error(`[thor-provider]Time out for whatever reason, check your provider`));
        };
        try {
            xhr.send(method === Method.POST ? JSON.stringify(body) : null);
        }
        catch (e) {
            return reject(new Error(`[thor-provider]Connect error: ${e.message}`));
        }
    });
};
