// @tatumio/tatum/dist/src/connector/tatum.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateBody = exports.httpDelete = exports.put = exports.postMultiForm = exports.post = exports.get = exports.axios = void 0;
const axios_1 = __importDefault(require("axios"));
const axios_retry_1 = __importStar(require("axios-retry"));
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const constants_1 = require("../constants");
exports.axios = axios_1.default.create({
    httpAgent: new http_1.default.Agent({ keepAlive: true }),
    httpsAgent: new https_1.default.Agent({ keepAlive: true })
});
// In case of 429 Too Many Requests response error, request is triggered again
axios_retry_1.default(exports.axios, {
    retryDelay: () => process.env.TATUM_RETRY_DELAY ? Number(process.env.TATUM_RETRY_DELAY) : constants_1.TATUM_RETRY_DELAY,
    retries: process.env.TATUM_RETRIES ? Number(process.env.TATUM_RETRIES) : constants_1.TATUM_RETRIES,
    retryCondition: (error) => { var _a; return axios_retry_1.isNetworkOrIdempotentRequestError(error) || ((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.status) === 429; }
});
const baseUrl = () => process.env.TATUM_API_URL || constants_1.TATUM_API_URL;
const headers = () => ({ headers: { 'x-api-key': process.env.TATUM_API_KEY, 'x-testnet-type': process.env.TESTNET_TYPE || 'ethereum-ropsten' } });
const get = async (url) => {
    const { data } = await exports.axios.get(`${baseUrl()}${url}`, headers());
    return data;
};
exports.get = get;
// eslint-disable-next-line @typescript-eslint/ban-types
const post = async (url, body, classType) => {
    await exports.validateBody(body, classType);
    const { data } = await exports.axios.post(`${baseUrl()}${url}`, body, headers());
    return data;
};
exports.post = post;
// eslint-disable-next-line @typescript-eslint/ban-types
const postMultiForm = async (url, body) => {
    const h = headers();
    h.headers = Object.assign(Object.assign({}, h.headers), body.getHeaders());
    const { data } = await exports.axios.post(`${baseUrl()}${url}`, body, h);
    return data;
};
exports.postMultiForm = postMultiForm;
// eslint-disable-next-line @typescript-eslint/ban-types
const put = async (url, body, classType) => {
    await exports.validateBody(body, classType);
    const { data } = await exports.axios.put(`${baseUrl()}${url}`, body, headers());
    return data;
};
exports.put = put;
const httpDelete = async (url) => {
    await exports.axios.delete(`${baseUrl()}${url}`, headers());
};
exports.httpDelete = httpDelete;
// eslint-disable-next-line @typescript-eslint/ban-types
const validateBody = async (body, classType) => {
    if (classType) {
        const classInstance = class_transformer_1.plainToClass(classType, body);
        await class_validator_1.validateOrReject(classInstance);
    }
};
exports.validateBody = validateBody;
