// @tatumio/tatum/dist/src/transaction/bitcoin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendBitcoinTransaction = exports.prepareBitcoinSignedTransaction = exports.signBitcoinKMSTransaction = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
// @ts-ignore
const bitcore_lib_1 = require("bitcore-lib");
const blockchain_1 = require("../blockchain");
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * Prepare a signed Btc transaction with the private key locally. Nothing is broadcasted to the blockchain.
 * @returns raw transaction data in hex, to be broadcasted to blockchain.
 */
const prepareSignedTransaction = async (body) => {
    var _a, _b;
    await tatum_1.validateBody(body, model_1.TransferBtcBasedBlockchain);
    const { fromUTXO, fromAddress, to } = body;
    const tx = new bitcore_lib_1.Transaction();
    const privateKeysToSign = [];
    if (fromAddress) {
        for (const item of fromAddress) {
            const txs = await blockchain_1.btcGetTxForAccount(item.address);
            for (const t of txs) {
                for (const [i, o] of t.outputs.entries()) {
                    if (o.address !== item.address) {
                        continue;
                    }
                    try {
                        await blockchain_1.btcGetUTXO(t.hash, i);
                        tx.from({
                            txId: t.hash,
                            outputIndex: i,
                            script: bitcore_lib_1.Script.fromAddress(item.address).toString(),
                            satoshis: o.value,
                        });
                        privateKeysToSign.push(item.signatureId || item.privateKey);
                    }
                    catch (e) {
                        console.error(e.toString());
                    }
                }
            }
        }
    }
    else if (fromUTXO) {
        for (const item of fromUTXO) {
            const t = await blockchain_1.btcGetTransaction(item.txHash);
            const address = t.outputs ? t.outputs[item.index].address : (_a = t.vout) === null || _a === void 0 ? void 0 : _a[item.index].scriptPubKey.addresses[0];
            const value = t.outputs ? t.outputs[item.index].value :
                Number(new bignumber_js_1.default(((_b = t.vout) === null || _b === void 0 ? void 0 : _b[item.index].value) || 0).multipliedBy(100000000).toFixed(8, bignumber_js_1.default.ROUND_FLOOR));
            tx.from({
                txId: item.txHash,
                outputIndex: item.index,
                script: bitcore_lib_1.Script.fromAddress(address).toString(),
                satoshis: value,
            });
            privateKeysToSign.push(item.signatureId || item.privateKey);
        }
    }
    for (const item of to) {
        tx.to(item.address, Number(new bignumber_js_1.default(item.value).multipliedBy(100000000).toFixed(8, bignumber_js_1.default.ROUND_FLOOR)));
    }
    if ((fromAddress && fromAddress[0].signatureId) || (fromUTXO && fromUTXO[0].signatureId)) {
        return JSON.stringify({ txData: JSON.stringify(tx), privateKeysToSign });
    }
    for (const item of privateKeysToSign) {
        tx.sign(bitcore_lib_1.PrivateKey.fromWIF(item));
    }
    return tx.serialize(true);
};
/**
 * Sign Bitcoin pending transaction from Tatum KMS
 * @param tx pending transaction from KMS
 * @param privateKeys private keys to sign transaction with.
 * @param testnet mainnet or testnet version
 * @returns transaction data to be broadcast to blockchain.
 */
const signBitcoinKMSTransaction = async (tx, privateKeys) => {
    if (tx.chain !== model_1.Currency.BTC) {
        throw Error('Unsupported chain.');
    }
    const builder = new bitcore_lib_1.Transaction(JSON.parse(tx.serializedTransaction));
    for (const privateKey of privateKeys) {
        builder.sign(bitcore_lib_1.PrivateKey.fromWIF(privateKey));
    }
    return builder.serialize(true);
};
exports.signBitcoinKMSTransaction = signBitcoinKMSTransaction;
/**
 * Sign Bitcoin transaction with private keys locally. Nothing is broadcast to the blockchain.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction data to be broadcast to blockchain.
 */
const prepareBitcoinSignedTransaction = async (testnet, body) => {
    return prepareSignedTransaction(body);
};
exports.prepareBitcoinSignedTransaction = prepareBitcoinSignedTransaction;
/**
 * Send Bitcoin transaction to the blockchain. This method broadcasts signed transaction to the blockchain.
 * This operation is irreversible.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction id of the transaction in the blockchain
 */
const sendBitcoinTransaction = async (testnet, body) => {
    return blockchain_1.btcBroadcast(await exports.prepareBitcoinSignedTransaction(testnet, body));
};
exports.sendBitcoinTransaction = sendBitcoinTransaction;
