// @tatumio/tatum/dist/src/transaction/xlm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareXlmSignedTransaction = exports.signXlmKMSTransaction = exports.sendXlmTransaction = void 0;
const stellar_sdk_1 = require("stellar-sdk");
const blockchain_1 = require("../blockchain");
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * Send Stellar transaction to the blockchain. This method broadcasts signed transaction to the blockchain.
 * This operation is irreversible.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction id of the transaction in the blockchain
 */
const sendXlmTransaction = async (testnet, body) => {
    return blockchain_1.xlmBroadcast(await exports.prepareXlmSignedTransaction(testnet, body));
};
exports.sendXlmTransaction = sendXlmTransaction;
/**
 * Sign Stellar pending transaction from Tatum KMS
 * @param tx pending transaction from KMS
 * @param secret secret key to sign transaction with.
 * @param testnet mainnet or testnet version
 * @returns transaction data to be broadcast to blockchain.
 */
const signXlmKMSTransaction = async (tx, secret, testnet) => {
    if (tx.chain !== model_1.Currency.XLM) {
        throw Error('Unsupported chain.');
    }
    const transaction = stellar_sdk_1.TransactionBuilder.fromXDR(tx.serializedTransaction, testnet ? stellar_sdk_1.Networks.TESTNET : stellar_sdk_1.Networks.PUBLIC);
    transaction.sign(stellar_sdk_1.Keypair.fromSecret(secret));
    return transaction.toEnvelope().toXDR().toString('base64');
};
exports.signXlmKMSTransaction = signXlmKMSTransaction;
/**
 * Sign Stellar transaction with private keys locally. Nothing is broadcast to the blockchain.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction data to be broadcast to blockchain.
 */
const prepareXlmSignedTransaction = async (testnet, body) => {
    await tatum_1.validateBody(body, model_1.TransferXlm);
    const { fromSecret, to, amount, message, initialize, } = body;
    const memo = message ? message.length > 28 ? stellar_sdk_1.Memo.hash(message) : stellar_sdk_1.Memo.text(message) : undefined;
    const fromAccount = stellar_sdk_1.Keypair.fromSecret(fromSecret).publicKey();
    const account = await blockchain_1.xlmGetAccountInfo(fromAccount);
    const builder = new stellar_sdk_1.TransactionBuilder(new stellar_sdk_1.Account(fromAccount, account.sequence), {
        fee: '100',
        networkPassphrase: testnet ? stellar_sdk_1.Networks.TESTNET : stellar_sdk_1.Networks.PUBLIC,
        memo,
    }).setTimeout(300);
    const tx = initialize
        ? builder.addOperation(stellar_sdk_1.Operation.createAccount({
            destination: to.trim(),
            startingBalance: amount,
        })).build()
        : builder.addOperation(stellar_sdk_1.Operation.payment({
            destination: to.trim(),
            asset: stellar_sdk_1.Asset.native(),
            amount,
        }))
            .build();
    tx.sign(stellar_sdk_1.Keypair.fromSecret(fromSecret));
    return tx.toEnvelope().toXDR().toString('base64');
};
exports.prepareXlmSignedTransaction = prepareXlmSignedTransaction;
// TODO: add support for TrustLine
