// thorify/dist/extend/contracts.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.extendContracts = void 0;
const debug = require('debug')('thor:injector');
const Subscription = require('web3-core-subscriptions').subscription;
const utils = require("../utils");
const formatters_1 = require("./formatters");
const extendContracts = function (web3) {
    const _encodeEventABI = web3.eth.Contract.prototype._encodeEventABI;
    web3.eth.Contract.prototype._encodeEventABI = function (event, options) {
        debug('_encodeEventABI');
        const result = _encodeEventABI.call(this, event, options);
        if (options.options) {
            result.options = options.options;
        }
        if (options.range) {
            result.range = options.range;
        }
        if (options.order) {
            result.order = options.order;
        }
        return result;
    };
    web3.eth.Contract.prototype._on = function () {
        debug('_on');
        // keeps the code from web3
        const subOptions = this._generateEventOptions.apply(this, arguments);
        // prevent the event "newListener" and "removeListener" from being overwritten
        this._checkListener('newListener', subOptions.event.name, subOptions.callback);
        this._checkListener('removeListener', subOptions.event.name, subOptions.callback);
        // TODO check if listener already exists? and reuse subscription if options are the same.
        // create new subscription
        /* changes from web3 starts the following
         */
        // subscription options in thor doesn't support array as topic filter object
        const filterOptions = {
            address: subOptions.params.address,
        };
        if (subOptions.params.fromBlock) {
            filterOptions.pos = utils.validBytes32OrError(subOptions.params.fromBlock, '[thorify] fromBlock must be blockID');
        }
        debug('Contract filter option: %O', filterOptions);
        if (subOptions.params.topics) {
            for (const [index, value] of subOptions.params.topics.entries()) {
                if (value === null) {
                    continue;
                }
                if (typeof value === 'string') {
                    filterOptions['t' + index] = value;
                }
                else {
                    throw new Error('[thorify] Array filter option is not supported in thor, must be null or bytes32 string');
                }
            }
        }
        const decodeEventABI = this._decodeEventABI.bind(subOptions.event);
        const subscription = new Subscription({
            subscription: {
                params: 1,
                inputFormatter: [formatters_1.inputLogFilterFormatter],
                subscriptionHandler(subscriptionMsg) {
                    if (subscriptionMsg.error) {
                        this.emit('error', subscriptionMsg.error);
                        // web3-core-subscriptions/subscription sets a default value for this.callback
                        this.callback(subscriptionMsg.error, null, this);
                    }
                    else {
                        const result = decodeEventABI(subscriptionMsg.data);
                        if (result.removed) {
                            this.emit('changed', result);
                        }
                        else {
                            this.emit('data', result);
                        }
                        // web3-core-subscriptions/subscription sets a default value for this.callback
                        this.callback(null, result, this);
                    }
                },
            },
            type: 'eth',
            requestManager: this._requestManager,
        });
        subscription.subscribe('logs', filterOptions, subOptions.callback || function () { });
        return subscription;
    };
};
exports.extendContracts = extendContracts;
