// @tatumio/tatum/dist/src/offchain/egld.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendEgldOffchainTransaction = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const transaction_1 = require("../transaction");
const wallet_1 = require("../wallet");
const common_1 = require("./common");
const kms_1 = require("./kms");
/**
 * Send EGLD transaction from Tatum Ledger account to the blockchain. This method broadcasts signed transaction to the blockchain.
 * This operation is irreversible.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @param provider url of the EGLD Server to connect to. If not set, default public server will be used.
 * @returns transaction id of the transaction in the blockchain or id of the withdrawal, if it was not cancelled automatically
 */
const sendEgldOffchainTransaction = async (testnet, body, provider) => {
    if (body.signatureId) {
        return kms_1.offchainTransferEgldKMS(body);
    }
    await tatum_1.validateBody(body, model_1.EgldTransferOffchain);
    const { mnemonic, index, fromPrivateKey, gasLimit, gasPrice } = body, withdrawal = __rest(body, ["mnemonic", "index", "fromPrivateKey", "gasLimit", "gasPrice"]);
    const { value, receiver } = withdrawal;
    const fromPriv = mnemonic && index !== undefined ? await wallet_1.generatePrivateKeyFromMnemonic(model_1.Currency.EGLD, testnet, mnemonic, index) : fromPrivateKey;
    const fee = {
        gasLimit: `${gasLimit || '50000'}`,
        gasPrice: `${gasPrice || '1000000000'}`,
    };
    const txData = await transaction_1.prepareEgldSignedTransaction({
        amount: value,
        fromPrivateKey: fromPriv,
        fee,
        to: receiver
    }, provider);
    // @ts-ignore
    withdrawal.fee = new bignumber_js_1.default(fee.gasLimit).multipliedBy(fee.gasPrice).toString();
    const { id } = await common_1.offchainStoreWithdrawal(withdrawal);
    try {
        return Object.assign(Object.assign({}, await common_1.offchainBroadcast({ txData, withdrawalId: id, currency: model_1.Currency.EGLD })), { id });
    }
    catch (e) {
        console.error(e);
        try {
            await common_1.offchainCancelWithdrawal(id);
        }
        catch (e1) {
            console.log(e);
            return { id };
        }
    }
};
exports.sendEgldOffchainTransaction = sendEgldOffchainTransaction;
