// @tatumio/tatum/dist/src/offchain/tron.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendTronOffchainTransaction = void 0;
const tatum_1 = require("../connector/tatum");
const constants_1 = require("../constants");
const ledger_1 = require("../ledger");
const request_1 = require("../model/request");
const transaction_1 = require("../transaction");
const wallet_1 = require("../wallet");
const common_1 = require("./common");
const kms_1 = require("./kms");
/**
 * Send Tron transaction from Tatum Ledger account to the blockchain. This method broadcasts signed transaction to the blockchain.
 * This operation is irreversible.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction id of the transaction in the blockchain or id of the withdrawal, if it was not cancelled automatically
 */
const sendTronOffchainTransaction = async (testnet, body) => {
    if (body.signatureId) {
        return kms_1.offchainTransferTronKMS(body);
    }
    await tatum_1.validateBody(body, request_1.TransferTrxOffchain);
    const { mnemonic, index, fromPrivateKey } = body, withdrawal = __rest(body, ["mnemonic", "index", "fromPrivateKey"]);
    const { amount, address } = withdrawal;
    let fromPriv;
    if (mnemonic && index !== undefined) {
        fromPriv = mnemonic && index ? await wallet_1.generatePrivateKeyFromMnemonic(request_1.Currency.TRON, testnet, mnemonic, index) : fromPrivateKey;
    }
    else if (fromPrivateKey) {
        fromPriv = fromPrivateKey;
    }
    else {
        throw new Error('No mnemonic or private key is present.');
    }
    withdrawal.fee = withdrawal.fee || '2.5';
    const account = await ledger_1.getAccountById(withdrawal.senderAccountId);
    let txData;
    if (account.currency === request_1.Currency.TRON) {
        txData = await transaction_1.prepareTronSignedTransaction(testnet, { amount, fromPrivateKey: fromPriv, to: address });
    }
    else if (account.currency === request_1.Currency.USDT_TRON || account.currency === request_1.Currency.INRT_TRON) {
        txData = await transaction_1.prepareTronTrc20SignedTransaction(testnet, {
            amount,
            fromPrivateKey: fromPriv,
            to: address,
            tokenAddress: constants_1.CONTRACT_ADDRESSES[account.currency],
            feeLimit: parseFloat(withdrawal.fee),
        });
    }
    else {
        const vc = await ledger_1.getVirtualCurrencyByName(account.currency);
        if (vc.trcType === request_1.TrcType.TRC10) {
            txData = await transaction_1.prepareTronTrc10SignedTransaction(testnet, {
                amount,
                fromPrivateKey: fromPriv,
                to: address,
                tokenId: vc.erc20Address,
            }, vc.precision);
        }
        else if (vc.trcType === request_1.TrcType.TRC20) {
            txData = await transaction_1.prepareTronTrc20SignedTransaction(testnet, {
                amount,
                feeLimit: parseFloat(withdrawal.fee),
                fromPrivateKey: fromPriv,
                to: address,
                tokenAddress: vc.erc20Address
            });
        }
        else {
            throw new Error('Unsupported account.');
        }
    }
    const { id } = await common_1.offchainStoreWithdrawal(withdrawal);
    try {
        return Object.assign(Object.assign({}, await common_1.offchainBroadcast({ txData, withdrawalId: id, currency: request_1.Currency.TRON })), { id });
    }
    catch (e) {
        console.error(e);
        try {
            await common_1.offchainCancelWithdrawal(id);
        }
        catch (e1) {
            console.log(e);
            return { id };
        }
    }
};
exports.sendTronOffchainTransaction = sendTronOffchainTransaction;
