// @tatumio/tatum/dist/src/offchain/xlm.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareXlmSignedOffchainTransaction = exports.signXlmOffchainKMSTransaction = exports.sendXlmOffchainTransaction = void 0;
const stellar_sdk_1 = require("stellar-sdk");
const blockchain_1 = require("../blockchain");
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const common_1 = require("./common");
/**
 * Send Stellar transaction from Tatum Ledger account to the blockchain. This method broadcasts signed transaction to the blockchain.
 * This operation is irreversible.
 * @param testnet mainnet or testnet version
 * @param body content of the transaction to broadcast
 * @returns transaction id of the transaction in the blockchain or id of the withdrawal, if it was not cancelled automatically
 */
const sendXlmOffchainTransaction = async (testnet, body) => {
    await tatum_1.validateBody(body, model_1.TransferXlmOffchain);
    const { secret } = body, withdrawal = __rest(body, ["secret"]);
    if (!withdrawal.fee) {
        withdrawal.fee = '0.00001';
    }
    const memo = withdrawal.attr ? withdrawal.attr.length > 28 ? stellar_sdk_1.Memo.hash(withdrawal.attr) : stellar_sdk_1.Memo.text(withdrawal.attr) : undefined;
    const account = await blockchain_1.xlmGetAccountInfo(stellar_sdk_1.Keypair.fromSecret(secret).publicKey());
    const { id } = await common_1.offchainStoreWithdrawal(withdrawal);
    const { amount, address, } = withdrawal;
    let txData;
    try {
        txData = await exports.prepareXlmSignedOffchainTransaction(testnet, account, amount, address, secret, memo);
    }
    catch (e) {
        console.error(e);
        await common_1.offchainCancelWithdrawal(id);
        throw e;
    }
    try {
        return Object.assign(Object.assign({}, await common_1.offchainBroadcast({ txData, withdrawalId: id, currency: model_1.Currency.XLM })), { id });
    }
    catch (e) {
        console.error(e);
        try {
            await common_1.offchainCancelWithdrawal(id);
        }
        catch (e1) {
            console.log(e);
            return { id };
        }
        throw e;
    }
};
exports.sendXlmOffchainTransaction = sendXlmOffchainTransaction;
/**
 * Sign Stellar pending transaction from Tatum KMS
 * @param tx pending transaction from KMS
 * @param secret secret key to sign transaction with.
 * @param testnet mainnet or testnet version
 * @returns transaction data to be broadcast to blockchain.
 */
const signXlmOffchainKMSTransaction = async (tx, secret, testnet) => {
    if (tx.chain !== model_1.Currency.XLM) {
        throw Error('Unsupported chain.');
    }
    const transaction = stellar_sdk_1.TransactionBuilder.fromXDR(tx.serializedTransaction, testnet ? stellar_sdk_1.Networks.TESTNET : stellar_sdk_1.Networks.PUBLIC);
    transaction.sign(stellar_sdk_1.Keypair.fromSecret(secret));
    return transaction.toEnvelope().toXDR().toString('base64');
};
exports.signXlmOffchainKMSTransaction = signXlmOffchainKMSTransaction;
/**
 * Sign Stellar transaction with private keys locally. Nothing is broadcast to the blockchain.
 * @param testnet mainnet or testnet version
 * @param account Stellar account with information
 * @param amount amount to send
 * @param address recipient address
 * @param secret secret to sign transaction with
 * @param memo short memo to include in transaction
 * @returns transaction data to be broadcast to blockchain.
 */
const prepareXlmSignedOffchainTransaction = async (testnet, account, amount, address, secret, memo) => {
    const builder = new stellar_sdk_1.TransactionBuilder(new stellar_sdk_1.Account(account.account_id, account.sequence), {
        fee: '100',
        networkPassphrase: testnet ? stellar_sdk_1.Networks.TESTNET : stellar_sdk_1.Networks.PUBLIC,
        memo,
    }).setTimeout(300);
    const tx = builder.addOperation(stellar_sdk_1.Operation.payment({
        destination: address,
        asset: stellar_sdk_1.Asset.native(),
        amount,
    })).build();
    tx.sign(stellar_sdk_1.Keypair.fromSecret(secret));
    return tx.toEnvelope().toXDR().toString('base64');
};
exports.prepareXlmSignedOffchainTransaction = prepareXlmSignedOffchainTransaction;
